<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" xml:lang="en-US">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="Robert Crystal-Ornelas" />
  <meta name="author" content="Brandon P.M. Edwards" />
  <meta name="author" content="Katherine Hébert" />
  <meta name="author" content="Emma J. Hudgins" />
  <meta name="author" content="Luna L. Sánchez Reyes" />
  <meta name="author" content="Eric R. Scott" />
  <meta name="author" content="Matthew J. Grainger" />
  <meta name="author" content="Vivienne Foroughirad" />
  <meta name="author" content="Allison D. Binley" />
  <meta name="author" content="Cole B. Brookson" />
  <meta name="author" content="Kaitlyn M. Gaynor" />
  <meta name="author" content="Saeed Shafiei Sabet" />
  <meta name="author" content="Ali Güncan" />
  <meta name="author" content="Friederike Hillemann" />
  <meta name="author" content="Helen Weierbach" />
  <meta name="author" content="Dylan G. E. Gomes" />
  <meta name="author" content="Pedro Henrique Pereira Braga" />
  <meta name="dcterms.date" content="2022-11-17" />
  <meta name="keywords" content="GitHub, Reproducible research, Collaboration, Ecology and evolution" />
  <title>Not just for programmers: How GitHub can accelerate collaborative and reproducible research in ecology and evolution</title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
    }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
  <!--
  Manubot generated metadata rendered from header-includes-template.html.
  Suggest improvements at https://github.com/manubot/manubot/blob/main/manubot/process/header-includes-template.html
  -->
  <meta name="dc.format" content="text/html" />
  <meta name="dc.title" content="Not just for programmers: How GitHub can accelerate collaborative and reproducible research in ecology and evolution" />
  <meta name="citation_title" content="Not just for programmers: How GitHub can accelerate collaborative and reproducible research in ecology and evolution" />
  <meta property="og:title" content="Not just for programmers: How GitHub can accelerate collaborative and reproducible research in ecology and evolution" />
  <meta property="twitter:title" content="Not just for programmers: How GitHub can accelerate collaborative and reproducible research in ecology and evolution" />
  <meta name="dc.date" content="2022-11-17" />
  <meta name="citation_publication_date" content="2022-11-17" />
  <meta name="dc.language" content="en-US" />
  <meta name="citation_language" content="en-US" />
  <meta name="dc.relation.ispartof" content="Manubot" />
  <meta name="dc.publisher" content="Manubot" />
  <meta name="citation_journal_title" content="Manubot" />
  <meta name="citation_technical_report_institution" content="Manubot" />
  <meta name="citation_author" content="Robert Crystal-Ornelas" />
  <meta name="citation_author_institution" content="Earth and Environmental Sciences Area, Lawrence Berkeley National Laboratory, Berkeley, CA 94720, USA" />
  <meta name="citation_author_orcid" content="0000-0002-6339-1139" />
  <meta name="twitter:creator" content="@rob_c_ornelas" />
  <meta name="citation_author" content="Brandon P.M. Edwards" />
  <meta name="citation_author_institution" content="Department of Biology, Carleton University, Ottawa, ON K1S 5B6, Canada" />
  <meta name="citation_author_orcid" content="0000-0003-0865-3076" />
  <meta name="citation_author" content="Katherine Hébert" />
  <meta name="citation_author_institution" content="Département de biologie, Université de Sherbrooke, Québec, Canada" />
  <meta name="citation_author_orcid" content="0000-0001-7866-6775" />
  <meta name="twitter:creator" content="@hebert_kat" />
  <meta name="citation_author" content="Emma J. Hudgins" />
  <meta name="citation_author_institution" content="Department of Biology, Carleton University, Ottawa, ON K1S 5B6, Canada" />
  <meta name="citation_author_orcid" content="0000-0002-8402-5111" />
  <meta name="citation_author" content="Luna L. Sánchez Reyes" />
  <meta name="citation_author_institution" content="School of Natural Sciences, University of California, Merced, USA" />
  <meta name="citation_author_orcid" content="0000-0001-7668-2528" />
  <meta name="twitter:creator" content="@LunaSare" />
  <meta name="citation_author" content="Eric R. Scott" />
  <meta name="citation_author_institution" content="Department of Wildlife Ecology and Conservation, University of Florida, Gainesville, FL, USA" />
  <meta name="citation_author_orcid" content="0000-0002-7430-7879" />
  <meta name="citation_author" content="Matthew J. Grainger" />
  <meta name="citation_author_institution" content="Terrestrial Biodiversity, Norwegian Institute for Nature Research - NINA, Postbox 5685 Torgarden, 7485 Trondheim, Norway" />
  <meta name="citation_author_orcid" content="0000-0001-8426-6495" />
  <meta name="twitter:creator" content="@Ed_pheasant" />
  <meta name="citation_author" content="Vivienne Foroughirad" />
  <meta name="citation_author_institution" content="Department of Biology, Georgetown University, Washington, DC, USA" />
  <meta name="citation_author_orcid" content="0000-0002-8656-7440" />
  <meta name="twitter:creator" content="@vforoughirad" />
  <meta name="citation_author" content="Allison D. Binley" />
  <meta name="citation_author_institution" content="Department of Biology, Carleton University, Ottawa, ON K1S 5B6, Canada" />
  <meta name="citation_author_orcid" content="0000-0001-8790-9935" />
  <meta name="twitter:creator" content="@AllisonBinley" />
  <meta name="citation_author" content="Cole B. Brookson" />
  <meta name="citation_author_institution" content="Department of Biological Sciences, University of Alberta, Edmonton, AB, Canada" />
  <meta name="citation_author_orcid" content="0000-0003-1237-4096" />
  <meta name="citation_author" content="Kaitlyn M. Gaynor" />
  <meta name="citation_author_institution" content="Departments of Zoology and Botany, University of British Columbia, Vancouver, BC, Canada" />
  <meta name="citation_author_institution" content="National Center for Ecological Analysis and Synthesis, Santa Barbara, CA 93101, USA" />
  <meta name="citation_author_orcid" content="0000-0002-5747-0543" />
  <meta name="twitter:creator" content="@kaitlyngaynor" />
  <meta name="citation_author" content="Saeed Shafiei Sabet" />
  <meta name="citation_author_institution" content="Fisheries Department, Faculty of Natural Resources, University of Guilan, Sowmeh Sara, Iran" />
  <meta name="citation_author_orcid" content="0000-0001-5919-2527" />
  <meta name="twitter:creator" content="@SaeedSHSABET" />
  <meta name="citation_author" content="Ali Güncan" />
  <meta name="citation_author_institution" content="Department of Plant Protection, Faculty of Agriculture, Ordu University, 52200, Ordu, Turkey" />
  <meta name="citation_author_orcid" content="0000-0003-1765-648X" />
  <meta name="twitter:creator" content="@aliguncan" />
  <meta name="citation_author" content="Friederike Hillemann" />
  <meta name="citation_author_institution" content="Department of Human Behavior, Ecology and Culture, Max Planck Institute for Evolutionary Anthropology, Leipzig, Germany" />
  <meta name="citation_author_orcid" content="0000-0002-8992-0676" />
  <meta name="citation_author" content="Helen Weierbach" />
  <meta name="citation_author_institution" content="Earth and Environmental Sciences Area, Lawrence Berkeley National Laboratory, Berkeley, CA 94720, USA" />
  <meta name="citation_author_orcid" content="0000-0001-6348-9120" />
  <meta name="twitter:creator" content="@HWeierbach" />
  <meta name="citation_author" content="Dylan G. E. Gomes" />
  <meta name="citation_author_institution" content="(Current) National Academy of Sciences NRC Research Associateship Program, Northwest Fisheries Science Center, National Marine Fisheries Service, National Oceanic and Atmospheric Administration, Seattle, WA, USA 98112" />
  <meta name="citation_author_institution" content="(Former) Cooperative Institute for Marine Resources Studies, Hatfield Marine Science Center, Oregon State University, Newport, OR, USA 97365" />
  <meta name="citation_author_orcid" content="0000-0002-2642-3728" />
  <meta name="citation_author" content="Pedro Henrique Pereira Braga" />
  <meta name="citation_author_institution" content="Department of Biology, Concordia University, 7141 Sherbrooke Street West, Montreal, QC H4B 1R6, Canada" />
  <meta name="citation_author_orcid" content="0000-0002-1308-1562" />
  <meta name="twitter:creator" content="@pedrohp_braga" />
  <link rel="canonical" href="https://SORTEE-Github-Hackathon.github.io/manuscript/" />
  <meta property="og:url" content="https://SORTEE-Github-Hackathon.github.io/manuscript/" />
  <meta property="twitter:url" content="https://SORTEE-Github-Hackathon.github.io/manuscript/" />
  <meta name="citation_fulltext_html_url" content="https://SORTEE-Github-Hackathon.github.io/manuscript/" />
  <meta name="citation_pdf_url" content="https://SORTEE-Github-Hackathon.github.io/manuscript/manuscript.pdf" />
  <link rel="alternate" type="application/pdf" href="https://SORTEE-Github-Hackathon.github.io/manuscript/manuscript.pdf" />
  <link rel="alternate" type="text/html" href="https://SORTEE-Github-Hackathon.github.io/manuscript/v/366399c3a3d8d06832f71ea5ec300bf690d17bdd/" />
  <meta name="manubot_html_url_versioned" content="https://SORTEE-Github-Hackathon.github.io/manuscript/v/366399c3a3d8d06832f71ea5ec300bf690d17bdd/" />
  <meta name="manubot_pdf_url_versioned" content="https://SORTEE-Github-Hackathon.github.io/manuscript/v/366399c3a3d8d06832f71ea5ec300bf690d17bdd/manuscript.pdf" />
  <meta property="og:type" content="article" />
  <meta property="twitter:card" content="summary_large_image" />
  <link rel="icon" type="image/png" sizes="192x192" href="https://manubot.org/favicon-192x192.png" />
  <link rel="mask-icon" href="https://manubot.org/safari-pinned-tab.svg" color="#ad1457" />
  <meta name="theme-color" content="#ad1457" />
  <!-- end Manubot generated metadata -->
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<header id="title-block-header">
<h1 class="title">Not just for programmers: How GitHub can accelerate collaborative and reproducible research in ecology and evolution</h1>
</header>
<p><small><em>
This manuscript
(<a href="https://SORTEE-Github-Hackathon.github.io/manuscript/v/366399c3a3d8d06832f71ea5ec300bf690d17bdd/">permalink</a>)
was automatically generated
from <a href="https://github.com/SORTEE-Github-Hackathon/manuscript/tree/366399c3a3d8d06832f71ea5ec300bf690d17bdd">SORTEE-Github-Hackathon/manuscript@366399c</a>
on November 17, 2022.
</em></small></p>
<h2 id="authors">Authors</h2>
<ul>
<li><p><strong>Robert Crystal-Ornelas</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-6339-1139">0000-0002-6339-1139</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/robcrystalornelas">robcrystalornelas</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/rob_c_ornelas">rob_c_ornelas</a><br>
<small>
Earth and Environmental Sciences Area, Lawrence Berkeley National Laboratory, Berkeley, CA 94720, USA
</small></p></li>
<li><p><strong>Brandon P.M. Edwards</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-0865-3076">0000-0003-0865-3076</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/BrandonEdwards">BrandonEdwards</a><br>
<small>
Department of Biology, Carleton University, Ottawa, ON K1S 5B6, Canada
</small></p></li>
<li><p><strong>Katherine Hébert</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-7866-6775">0000-0001-7866-6775</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/katherinehebert">katherinehebert</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/hebert_kat">hebert_kat</a><br>
<small>
Département de biologie, Université de Sherbrooke, Québec, Canada
</small></p></li>
<li><p><strong>Emma J. Hudgins</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-8402-5111">0000-0002-8402-5111</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/emmajhudgins">emmajhudgins</a><br>
<small>
Department of Biology, Carleton University, Ottawa, ON K1S 5B6, Canada
</small></p></li>
<li><p><strong>Luna L. Sánchez Reyes</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-7668-2528">0000-0001-7668-2528</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/LunaSare">LunaSare</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/LunaSare">LunaSare</a><br>
<small>
School of Natural Sciences, University of California, Merced, USA
</small></p></li>
<li><p><strong>Eric R. Scott</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-7430-7879">0000-0002-7430-7879</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/Aariq">Aariq</a><br>
<small>
Department of Wildlife Ecology and Conservation, University of Florida, Gainesville, FL, USA
</small></p></li>
<li><p><strong>Matthew J. Grainger</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-8426-6495">0000-0001-8426-6495</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/DrMattG">DrMattG</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/Ed_pheasant">Ed_pheasant</a><br>
<small>
Terrestrial Biodiversity, Norwegian Institute for Nature Research - NINA, Postbox 5685 Torgarden, 7485 Trondheim, Norway
</small></p></li>
<li><p><strong>Vivienne Foroughirad</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-8656-7440">0000-0002-8656-7440</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/vjf2">vjf2</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/vforoughirad">vforoughirad</a><br>
<small>
Department of Biology, Georgetown University, Washington, DC, USA
</small></p></li>
<li><p><strong>Allison D. Binley</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-8790-9935">0000-0001-8790-9935</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/adbinley">adbinley</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/AllisonBinley">AllisonBinley</a><br>
<small>
Department of Biology, Carleton University, Ottawa, ON K1S 5B6, Canada
</small></p></li>
<li><p><strong>Cole B. Brookson</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-1237-4096">0000-0003-1237-4096</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/colebrookson">colebrookson</a><br>
<small>
Department of Biological Sciences, University of Alberta, Edmonton, AB, Canada
</small></p></li>
<li><p><strong>Kaitlyn M. Gaynor</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-5747-0543">0000-0002-5747-0543</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/kaitlyngaynor">kaitlyngaynor</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/kaitlyngaynor">kaitlyngaynor</a><br>
<small>
Departments of Zoology and Botany, University of British Columbia, Vancouver, BC, Canada; National Center for Ecological Analysis and Synthesis, Santa Barbara, CA 93101, USA
</small></p></li>
<li><p><strong>Saeed Shafiei Sabet</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-5919-2527">0000-0001-5919-2527</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/shafieisabets">shafieisabets</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/SaeedSHSABET">SaeedSHSABET</a><br>
<small>
Fisheries Department, Faculty of Natural Resources, University of Guilan, Sowmeh Sara, Iran
</small></p></li>
<li><p><strong>Ali Güncan</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-1765-648X">0000-0003-1765-648X</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/Aguncan">Aguncan</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/aliguncan">aliguncan</a><br>
<small>
Department of Plant Protection, Faculty of Agriculture, Ordu University, 52200, Ordu, Turkey
</small></p></li>
<li><p><strong>Friederike Hillemann</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-8992-0676">0000-0002-8992-0676</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/fhillemann">fhillemann</a><br>
<small>
Department of Human Behavior, Ecology and Culture, Max Planck Institute for Evolutionary Anthropology, Leipzig, Germany
</small></p></li>
<li><p><strong>Helen Weierbach</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-6348-9120">0000-0001-6348-9120</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/helenweierbach">helenweierbach</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/HWeierbach">HWeierbach</a><br>
<small>
Earth and Environmental Sciences Area, Lawrence Berkeley National Laboratory, Berkeley, CA 94720, USA
</small></p></li>
<li><p><strong>Dylan G. E. Gomes</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-2642-3728">0000-0002-2642-3728</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/dylangomes">dylangomes</a><br>
<small>
(Current) National Academy of Sciences NRC Research Associateship Program, Northwest Fisheries Science Center, National Marine Fisheries Service, National Oceanic and Atmospheric Administration, Seattle, WA, USA 98112; (Former) Cooperative Institute for Marine Resources Studies, Hatfield Marine Science Center, Oregon State University, Newport, OR, USA 97365
</small></p></li>
<li><p><strong>Pedro Henrique Pereira Braga</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-1308-1562">0000-0002-1308-1562</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/pedrohbraga">pedrohbraga</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/pedrohp_braga">pedrohp_braga</a><br>
<small>
Department of Biology, Concordia University, 7141 Sherbrooke Street West, Montreal, QC H4B 1R6, Canada
</small></p></li>
</ul>
<h2 class="page_break_before" id="abstract">Abstract</h2>
<p>Researchers in ecology and evolutionary biology are increasingly dependent on computational code to conduct research, and the use of efficient methods to share, reproduce, and collaborate on code as well as any research-related documentation has become fundamental.
GitHub is an online, cloud-based service that can help researchers track, organize, discuss, share, and collaborate on software and other materials related to research production, including data, code for analyses, and protocols.</p>
<p>Despite these benefits, the use of GitHub in ecology and evolution is not widespread.
To help researchers in ecology and evolution adopt useful features from GitHub in their own workflows, we review thirteen practical ways to use the platform.
We outline features ranging from low to high technical difficulty: storing code, managing projects, coding collaboratively, conducting peer review, and writing a manuscript.
Given that members of a research team may have different technical skills and responsibilities, we describe how the optimal use of GitHub features may vary among members of a research collaboration.
As more ecologists and evolutionary biologists establish their workflows using GitHub, the field can continue to push the boundaries of collaborative, transparent, and open research.</p>
<h2 class="page_break_before" id="introduction">Introduction</h2>
<!-- *Contributors to this section: PHPB* -->
<p>Most scientists, including ecologists and evolutionary biologists, are increasingly dependent on computational tools in their research <span class="citation" data-cites="fJWFe93e">(<a href="#ref-fJWFe93e" role="doc-biblioref">Hannay et al., 2009</a>)</span>.
Researchers write and use software packages or data analysis code (hereafter, code) to perform scientific tasks ranging from data management, data analysis, and study replication, to the application and the development of tools for hypothesis testing.
Maintaining code for scientific collaboration requires an efficient and well-documented work-flow <span class="citation" data-cites="1Kqna6l2">(<a href="#ref-1Kqna6l2" role="doc-biblioref">J. M. Perkel, 2020</a>)</span>.
To facilitate this process, scientists have been adopting tools from information and systems technology, such as cloud-based services for documentation and version control (<em>e.g.</em>, from the Google Suite, the Microsoft Suite, and GitHub <span class="citation" data-cites="10ghgV3S8">(<a href="#ref-10ghgV3S8" role="doc-biblioref">J. Perkel, 2016</a>)</span>).
However, most researchers lack exposure to adequate code practices and thus dedicate valuable time and effort to self-teaching research-facilitating tools.
Thus, researchers may not adhere to standards of code quality and maintenance <span class="citation" data-cites="10SpoByIw O6UbstGG fJWFe93e">(<a href="#ref-fJWFe93e" role="doc-biblioref">Hannay et al., 2009</a>; <a href="#ref-10SpoByIw" role="doc-biblioref">Prabhu et al., 2011</a>; <a href="#ref-O6UbstGG" role="doc-biblioref">Wilson et al., 2014</a>)</span>.
Here, we review and discuss one of the most used web-based platforms for computational version control and collaboration, GitHub, and provide researchers in ecology and evolutionary biology (EEB) with practical workflows to facilitate and improve their code and its management.</p>
<!-- *Contributors to this section: RCO, SSHS, PHPB* -->
<p>With over 83 million registered users as of 2022, GitHub (<a href="https://github.com" class="uri">https://github.com</a>) is the most widely-used web platform for collaborating on computer code <span class="citation" data-cites="nwCtHDCn">(<a href="#ref-nwCtHDCn" role="doc-biblioref"><em>Build Software Better, Together</em>, n.d.</a>)</span>.
GitHub provides a simple but powerful web interface that allows users to participate in projects by contributing, modifying and discussing existing code, reporting bugs, discovering code and data, and publishing new code.
Through version control, users have a detailed, chronological record of the files and directories stored in their repositories <span class="citation" data-cites="RVetqmsg">(<a href="#ref-RVetqmsg" role="doc-biblioref">Bryan, 2018</a>)</span> (see <a href="#definitions">Box 1</a>).
This workflow provides a strong and clear advantage over sending files back-and-forth (<em>e.g.</em>, via email), a process that can become challenging and time-consuming in long-term and collaborative projects <span class="citation" data-cites="4ny1onB0">(<a href="#ref-4ny1onB0" role="doc-biblioref">Ram, 2013</a>)</span>.
Through its combination of version control and collaborative features, GitHub facilitates open-source code alongside collaborative development <span class="citation" data-cites="kEX5dgzK">(<a href="#ref-kEX5dgzK" role="doc-biblioref">Perez-Riverol et al., 2016</a>)</span>.</p>
<p>Git is the version control system that enables all the collaborative tools available on GitHub.
Although the understanding of basic concepts of Git (such as commit, push, pull, checkout; see <a href="#definitions">Box 1</a>) is necessary, the GitHub web-based platform and its integrated development environments (such as the GitHub Desktop) allow users to perform most repository and data management operations without opening Git command-line sessions.</p>
<!-- *Contributors to this section: RCO, PHPB* -->
<p>The expansive GitHub user-community and numerous GitHub resources have boosted its popularity [<span class="citation" data-cites="kEX5dgzK">Perez-Riverol et al. (<a href="#ref-kEX5dgzK" role="doc-biblioref">2016</a>)</span>; <span class="citation" data-cites="RVetqmsg">Bryan (<a href="#ref-RVetqmsg" role="doc-biblioref">2018</a>)</span>; <a href="https://happygitwithr.com" class="uri">https://happygitwithr.com</a>; <a href="https://ourcodingclub.github.io" class="uri">https://ourcodingclub.github.io</a>].
Nevertheless, although multiple articles have encouraged researchers in EEB to adopt GitHub as part of their research process <span class="citation" data-cites="3DKwn1sY 10ghgV3S8">(<a href="#ref-3DKwn1sY" role="doc-biblioref">Lowndes et al., 2017</a>; <a href="#ref-10ghgV3S8" role="doc-biblioref">J. Perkel, 2016</a>)</span>, its use is still not widespread.
First-time users without formal training in information technology may face steep learning curves because GitHub and its features have been centered on collaboration for code development in information systems <span class="citation" data-cites="139b0pSGc">(<a href="#ref-139b0pSGc" role="doc-biblioref">Leibzon, 2016</a>)</span>.
Moreover, domain-specific resources providing tractable examples and practical guidance for researchers in EEB on GitHub are scarce [but see <span class="citation" data-cites="lJAgyhYq">Kim et al. (<a href="#ref-lJAgyhYq" role="doc-biblioref">2022</a>)</span>; <a href="https://ourcodingclub.github.io" class="uri">https://ourcodingclub.github.io</a>; <a href="https://www.openscapes.org" class="uri">https://www.openscapes.org</a>].
Widespread adoption of GitHub for collaborating on research tasks can ultimately enable EEB researchers to spend less time on creating novel processes for collaboration and more time on their research <span class="citation" data-cites="ydrk01SR">(<a href="#ref-ydrk01SR" role="doc-biblioref">Briney et al., 2020</a>)</span>.
More importantly, expanding the availability of data and code management standards – of which GitHub is one increasingly important component – makes research more reproducible and collaborative <span class="citation" data-cites="13QX8XU3J pq2Tv1BC">(<a href="#ref-13QX8XU3J" role="doc-biblioref">Alston &amp; Rick, 2021</a>; <a href="#ref-pq2Tv1BC" role="doc-biblioref">Gomes et al., 2022</a>)</span>.</p>
<!-- *Contributors to this section: RCO, PHPB* -->
<p>This paper is the result of an academic hackathon held during the 2021 conference for the Society for Open, Reliable, and Transparent Ecology and Evolutionary Biology (SORTEE, <a href="https://www.sortee.org" class="uri">https://www.sortee.org</a>).
We convened around thirty researchers in EEB with varying levels of familiarity with the usage of GitHub in research projects to showcase and discuss how existing features can contribute to documentation and collaboration in EEB research.
During the hackathon, we identified the need for formal discussions on how EEB researchers can benefit from GitHub and its features.
Here, we outline thirteen practical ways that EEB researchers can use GitHub features for more collaborative, transparent, and reproducible science.
We also provide critical perspectives on features that could be improved and catered towards research development.</p>
<h3 id="definitions">Box 1</h3>
<!-- Contributors to this section: ERS, Ali -->
<table>
<colgroup>
<col style="width: 100%" />
</colgroup>
<thead>
<tr class="header">
<th>Glossary</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>repository</strong>: A repository (commonly shortened to “repo”) is a collection of files (<em>e.g.</em>, a directory) tracked by Git. Repositories are managed by an owner and can be made either “public”, to be visible to all GitHub users, or “private”, to selected owner-specified users. Repositories can be either “local” and saved on an individual’s computer or “remote” and stored on the cloud via GitHub’s web platform.</td>
</tr>
<tr class="even">
<td><strong>fork</strong>: A fork is a copy of a <strong>repository</strong> hosted on GitHub. If a repository is public, then anyone can make a fork. Even if they do not have access to <strong>push</strong> to the original repository, they can make a fork and edit it independently. Forks are linked to the original GitHub repository and “upstream” changes (<em>i.e.</em>, those in the original repository) can be <strong>merged</strong> to keep the fork up-to-date with the original project. Changes made in the fork can be integrated into the original project via <strong>pull requests</strong>.</td>
</tr>
<tr class="odd">
<td><strong>clone</strong>: Cloning a <strong>repository</strong> is a way of making a local copy (<em>i.e.</em>, on your computer) of a GitHub <strong>repository</strong>. If you have access to <strong>push</strong> to a <strong>repository</strong>, this can be a first step to contributing to a project.</td>
</tr>
<tr class="even">
<td><strong>branch</strong>: Git workflow timelines or repositories are analogous to trees, with a main working project and diverging branches that are pointers to changes during the development process. A git branch is an alternative line of development for a project (repository). Branches allow users to add new features or modifications to the project without affecting the main part of the project. Development branches can be created at any point in time and work on each branch can continue independently. Branching is useful for testing out new ideas (both code and text) which may or may not eventually get integrated into the main branch of the project. Branches can also be used to isolate contributions of multiple contributors. Each person working on their own branch eliminates problems that may arise if conflicting edits are <strong>pushed</strong> to the same remote branch. Changes in a development branch can be <strong>merged</strong> into the main branch via <strong>pull requests</strong>. Branches can only be made by those who are given access to the project <strong>repository</strong>.</td>
</tr>
<tr class="odd">
<td><strong>commit</strong>: Commits are snapshots of the development of a project. In Git, versions of files and directories are uniquely identified as “commits”, allowing one to identify and track modifications line-by-line. Commits can include changes in multiple files and must include a brief commit message describing the changes made. A typical workflow is to make some related changes in files, add a commit message (<em>e.g.</em>, “Generate and include results figure”), and after several commits <strong>push</strong> those commits to the remote (<em>i.e.</em>, cloud-based) GitHub <strong>repository</strong>.</td>
</tr>
<tr class="even">
<td><strong>push and pull</strong>: When <strong>commits</strong> are made in a project locally, they must be synced with the remote GitHub repository by <strong>pushing</strong> them. Changes on a GitHub repository can then be <strong>pulled</strong> to keep your local version of the project up-to-date with the remote.</td>
</tr>
<tr class="odd">
<td><strong>pull request</strong>: A pull request is a request for changes made on an individual’s <strong>branch</strong> in the repository or in a user’s <strong>fork</strong> to be merged to the repository. Pull requests contain a description of the changes alongside all code required for testing and review by other users prior to being merged into the repository.</td>
</tr>
<tr class="even">
<td><strong>merge</strong>: Combining <strong>commits</strong> from two different branches together into one <strong>branch</strong>.</td>
</tr>
<tr class="odd">
<td><strong>release</strong>: At any point a release can be made on GitHub to mark a significant milestone in the progression of a <strong>repository</strong>. While this GitHub feature is designed with releases of new versions of code in mind (<em>e.g.</em>, v1.0.0), it can also be used to create a snapshot of a repository at significant stages like pre-print, submission, revision, and acceptance of an associated manuscript.</td>
</tr>
<tr class="even">
<td><strong>community</strong>: A forum where GitHub users can ask for advice, offer solutions to questions, and share ideas (<a href="https://github.community/" class="uri">https://github.community/</a>).</td>
</tr>
</tbody>
</table>
<h2 id="thirteen-practical-ways-github-can-accelerate-research-in-ecology-and-evolution">Thirteen practical ways GitHub can accelerate research in ecology and evolution</h2>
<h3 id="storing-and-sharing-research-compendia">Storing and sharing research compendia</h3>
<!--*contributors to this section: Dylan Gomes, Emma Hudgins, Pedro Braga* -->
<p>An EEB research compendium includes all computational materials related to research production, including data, code for analyses and protocols.
Having copies of these files safely stored is essential to protect against accidental modifications or deletions.
Many researchers begin using GitHub to store (or backup) their research compendium <span class="citation" data-cites="MwwMapRG">(<a href="#ref-MwwMapRG" role="doc-biblioref">Marwick et al., 2018</a>)</span> to a centralized, readily-available remote server (see <a href="#definitions">Box 1</a>).
A centralized research compendium stored in a version-controlled repository has the advantages of facilitating collaboration, integrating data and code archiving services, allowing file versions to be accessed and restored, and contributes to open science (see sections below).</p>
<p>GitHub limits committed file sizes to 100 Mb (megabytes) <span class="citation" data-cites="1Co6ZZjF1">(<a href="#ref-1Co6ZZjF1" role="doc-biblioref"><em>About Large Files on GitHub</em>, n.d.</a>)</span>, which can make it challenging for centralizing research compendia containing larger file sizes.
Users may still version large files using Git Large File Storage (<a href="https://git-lfs.github.com" class="uri">https://git-lfs.github.com</a>) text pointers, but may have to rely on external file storage alternatives (such as local or cloud-hosting).</p>
<h3 id="virtual-laboratory-notebooks">Virtual laboratory notebooks</h3>
<!--*contributors to this section: RCO* -->
<p>Laboratory notebooks help researchers track their research notes, methods, policies and protocols <span class="citation" data-cites="wwHxTOtm">(<a href="#ref-wwHxTOtm" role="doc-biblioref">Kanza et al., 2017</a>)</span>.
Virtual laboratory notebooks can be stored in GitHub repositories and provide the benefits of simultaneous, centralized and selective access, and allows for the easy update of policies and experiment protocols <span class="citation" data-cites="10V7x4H4l">(<a href="#ref-10V7x4H4l" role="doc-biblioref">Schnell, 2015</a>)</span>.
Researchers have been increasingly using GitHub to maintain versions and share digital laboratory notebooks across many fields <span class="citation" data-cites="10ghgV3S8">(<a href="#ref-10ghgV3S8" role="doc-biblioref">J. Perkel, 2016</a>)</span>, including applied ecology (<em>e.g.</em>, <a href="https://scheuerell-lab.github.io/lab-book" class="uri">https://scheuerell-lab.github.io/lab-book</a>), biogeography and global change biology (<em>e.g.</em>, <a href="https://github.com/HuckleyLab/how_we_work" class="uri">https://github.com/HuckleyLab/how_we_work</a>), and microbial ecology (<em>e.g.</em>, <a href="https://github.com/CarBBAS/uqam-guide" class="uri">https://github.com/CarBBAS/uqam-guide</a>).</p>
<p>At least for aspects of a research project that involve writing code, a GitHub repository is a form of a laboratory notebook; when changes are made to files in a version-controlled repository, the author of those changes makes a commit (<a href="#definitions">Box 1</a>) accompanied by a description of changes.
Later, the entire history of commits and their commit messages are viewable and can be audited similar to a physical laboratory notebook <span class="citation" data-cites="4ny1onB0">(<a href="#ref-4ny1onB0" role="doc-biblioref">Ram, 2013</a>)</span>.
GitHub issues (see Project management section below) can be used to prioritize laboratory objectives and goals, as well as track any progress updates.</p>
<h3 id="project-management">Project management</h3>
<!--*Contributors to this section: Kaitlyn Gaynor, Rob Crystal-Ornelas, Pedro Braga*-->
<p>Modern research in ecology and evolution is highly collaborative, bringing together multidisciplinary teams from various institutions <span class="citation" data-cites="1HhzKAC1K">(<a href="#ref-1HhzKAC1K" role="doc-biblioref">Goring et al., 2014</a>)</span>.
On GitHub, collaborators can share feedback, brainstorm ideas, and troubleshoot problems (Figure <a href="#fig:github-diagram">1</a>).
Project management can happen via three GitHub repository features: “Issues”, “Discussion” and “Projects”.
Github Issues allow for discrete tasks and sub-tasks to be identified, assigned to team members, and categorized with custom labels.
Github Discussions serve as a message board for conversation.
Finally, GitHub Projects integrate issues and pull requests on automated spreadsheets and project boards, providing users with real-time tracking of project priorities and status <span class="citation" data-cites="RhBKe0MG">(<a href="#ref-RhBKe0MG" role="doc-biblioref"><em>About Projects</em>, n.d.</a>)</span>.
Scripts, commit messages, and pull requests can be linked directly to issues, discussions, and projects providing a clear record of project workflow.
For instance, one can create a discussion on a project repository to decide which method to apply for biodiversity data analysis.
Then, an issue can be created to establish steps and responsibilities including data formatting, statistical analyses, figure generation, and issue resolution <span class="citation" data-cites="doi:10.1111/geb.13346">(<em>e.g.</em>, see issues and discussions in the sPlotOpen manuscript repository, <a href="#ref-doi:10.1111/geb.13346" role="doc-biblioref"><strong>doi:10.1111/geb.13346?</strong></a>)</span>; <span class="citation" data-cites="3UAritXO"><em>Issues · Fmsabatini/sPlotOpen_Manuscript</em> (<a href="#ref-3UAritXO" role="doc-biblioref">n.d.</a>)</span>=is%3Aissue+is%3Aclosed].
Using GitHub for all project-related conversation and planning, rather than email or messaging tools, makes it easier to keep track of progress throughout the lifespan of a project.
Unlike emails and messages which can get lost as more new tasks arise, GitHub issues are intentionally closed by repository administrators hiding the issue from view (closed issues remain accessible but not immediately visible).
Project management in GitHub can also be integrated with third-party applications, such as Zenhub (<a href="https://www.zenhub.com" class="uri">https://www.zenhub.com</a>) or Slack (<a href="https://slack.github.com" class="uri">https://slack.github.com</a>).
Zenhub allows for the enhanced visualization and organization of repositories and their issues, while the GitHub for Slack integration allows notifications from GitHub events to be sent directly to users or group channels <span class="citation" data-cites="1GRWGlDWy">(see <a href="#ref-1GRWGlDWy" role="doc-biblioref">Slack, n.d.</a>)</span>.</p>
<h3 id="educational-materials">Educational materials</h3>
<!-- *contributors to this section: Cole Brookson* -->
<p>GitHub supports a broad set of mechanisms for hosting educational materials.
The entire process of running a course, workshop, or lecture, can all be done openly on GitHub including material development, web hosting, and delivery, and even submission and grading of assignments.
While there are other purpose-built platforms for this, GitHub provides a free, open-source alternative.</p>
<p>Making presentations, syllabi and other course materials can be done through most major high-level programming languages such as R, with <code>RMarkdown</code> <span class="citation" data-cites="QqMezOMg">(<a href="#ref-QqMezOMg" role="doc-biblioref">Xie et al., 2019</a>)</span>, Python, with <code>python-ppt</code> (<a href="https://python-pptx.readthedocs.io" class="uri">https://python-pptx.readthedocs.io</a>), and Julia, with <code>Remark.jl</code> (<a href="https://juliapackages.com/p/remark" class="uri">https://juliapackages.com/p/remark</a>), and be version-controlled and stored in GitHub.
Once content is made, hosting a course website can be done through GitHub Pages (<a href="https://pages.github.com" class="uri">https://pages.github.com</a>; <em>e.g.</em>, <a href="https://github.com/topics/course-website" class="uri">https://github.com/topics/course-website</a>).
This way, course content can be available to enrolled students, as well as a global pool of learners and teachers interested in the material.
Content can then be delivered via the course website, and/or a GitHub Organization with, for example, template repositories for assignments.
Student submissions are perhaps the most challenging component, but the new GitHub Classroom tool (<a href="https://classroom.github.com" class="uri">https://classroom.github.com</a>) allows instructors to host private assignments to be submitted collaboratively or individually as code or PDF files, and even build autograding tests.
Although time-consuming to establish, using these features can integrate learning version control and GitHub with the learning course content, and thus boost students’ feelings of self-efficacy and confidence <span class="citation" data-cites="dqrFjoSb">(<a href="#ref-dqrFjoSb" role="doc-biblioref">Trujillo &amp; Tanner, 2014</a>)</span>.</p>
<!--## GitHub in EcoEvo examples (Part 2)-->
<h3 id="hosting-a-website">Hosting a website</h3>
<!--*Contributors to this section: Rob Crystal-Ornelas, Emma Hudgins*   -->
<p>Personal or laboratory websites can improve the sharing of research findings, build online presence, and increase coordination of research efforts <span class="citation" data-cites="HiIPSSHV">(<a href="#ref-HiIPSSHV" role="doc-biblioref">Smaglik, 2007</a>)</span>.
Despite researchers in ecology and evolution generally lacking experience in building or hosting webpages, many tools have been developed to help this process.
Static websites can now be easily built using independent software and languages, such as Jekyll (<a href="https://jekyllrb.com" class="uri">https://jekyllrb.com</a>), Hugo (<a href="https://gohugo.io" class="uri">https://gohugo.io</a>), Quarto (<a href="https://quarto.org" class="uri">https://quarto.org</a>), and wowchemy (<a href="https://wowchemy.com" class="uri">https://wowchemy.com</a>), or with the help of dependencies in the programming languages scientists commonly use, such as the <code>distill</code> (<a href="https://github.com/rstudio/distill" class="uri">https://github.com/rstudio/distill</a>) and <code>blogdown</code> (<a href="https://bookdown.org/yihui/blogdown" class="uri">https://bookdown.org/yihui/blogdown</a>) R packages.
The resulting files from static websites can then be hosted in repositories, from which one can activate GitHub’s Pages (<a href="https://pages.github.com" class="uri">https://pages.github.com</a>) feature, allowing for the direct live hosting of HyperText Markup Language (HTML), Cascading Style Sheets (CSS) and JavaScript files within a free github.io domain.</p>
<p>Website templates are readily available on GitHub (<a href="https://github.com/topics/website-template" class="uri">https://github.com/topics/website-template</a>) or in user-specific repositories, which can be forked (<a href="#definitions">Box 1</a>) and customized <span class="citation" data-cites="MXxgZJ45">(<a href="#ref-MXxgZJ45" role="doc-biblioref">Dawson, 2016</a>)</span>.
Aside from free hosting services, GitHub Pages also allows websites to be autogenerated and instantly updated following content modifications <span class="citation" data-cites="RVetqmsg">(<a href="#ref-RVetqmsg" role="doc-biblioref">Bryan, 2018</a>)</span>.
Creating and hosting websites on GitHub pages can be time-consuming, as this process usually requires more knowledge in web design than out-of-the-box platforms (<em>e.g.</em>, Wix, Weebly, Google Sites).
However, free hosting, widely available template customization, and versioning are strong advantages over the alternatives.</p>
<h3 id="archiving-citable-code-and-data">Archiving citable code and data</h3>
<!--*Contributors to this section: Rob Crystal-Ornelas, Emma Hudgins, Dylan Gomes, PHPB*-->
<p>Government, funding agencies, and publishers exercise rigorous open access data policies and mandates <span class="citation" data-cites="PLmDFZrm 1Hcf13Q0k">(<a href="#ref-1Hcf13Q0k" role="doc-biblioref">Nugroho et al., 2015</a>; <a href="#ref-PLmDFZrm" role="doc-biblioref">Tenopir et al., 2020</a>)</span>.
However, code and data sharing may be met by individual reluctance, temporary embargoes, or partially prevented by privacy and confidentiality reasons <span class="citation" data-cites="1Ch6LSHef SLq38RVv 1CzUZwyU2">(<a href="#ref-SLq38RVv" role="doc-biblioref">Figueiredo, 2017</a>; <a href="#ref-1CzUZwyU2" role="doc-biblioref">Tenopir et al., 2015</a>; <a href="#ref-1Ch6LSHef" role="doc-biblioref">Wicherts et al., 2011</a>)</span>.
Still, data deposition and ensuring its availability can amplify the outreach of published studies <span class="citation" data-cites="666HppfO">(<a href="#ref-666HppfO" role="doc-biblioref">Pronk et al., 2015</a>)</span>, increase citation rates <span class="citation" data-cites="1CcAUn3Lu">(<a href="#ref-1CcAUn3Lu" role="doc-biblioref">Piwowar et al., 2007</a>)</span>, and among many other reasons, enable the reproducibility and robustness of scientific advances <span class="citation" data-cites="4LaijDIZ uBJwnPbq 1HZdsK5Kn">(<a href="#ref-1HZdsK5Kn" role="doc-biblioref">Baker, 2016</a>; <a href="#ref-uBJwnPbq" role="doc-biblioref">Mislan et al., 2016</a>; <a href="#ref-4LaijDIZ" role="doc-biblioref"><span>“On Data Availability, Reproducibility and Reuse,”</span> 2017</a>)</span>.
While public repositories on GitHub make it easy to store and share data files, they are not considered long-term repositories for research materials.
This is because GitHub, a for-profit company, does not have long-term data availability guarantees, allowing users to delete or make repositories private after publication.
Also, GitHub does not issue Digital Object Identifiers (DOI) for content uploaded to their servers.
DOIs are persistent and citable unique alpha-numeric identifiers assigned to digitally stored research materials.
Because of this, scientists sharing code and data through GitHub are strongly encouraged to independently submit their research materials to long-term data archives (<em>e.g.</em>, Zenodo, Figshare, Dryad, OSF <span class="citation" data-cites="10ghgV3S8 kEX5dgzK 1Du6fzB8g">(<a href="#ref-1Du6fzB8g" role="doc-biblioref">Crystal‐Ornelas et al., 2021</a>; <a href="#ref-kEX5dgzK" role="doc-biblioref">Perez-Riverol et al., 2016</a>; <a href="#ref-10ghgV3S8" role="doc-biblioref">J. Perkel, 2016</a>)</span>; <a href="#tbl:compare">Table 1</a>).
Some of these options (Zenodo, Figshare and OSF) integrate with GitHub, allowing project, code, and data releases (<a href="#definitions">Box 1</a>) to be archived with versioned, citable DOIs.
Linking GitHub repositories with a DOI helps research become findable, properly cited, and can ensure long-term stability <span class="citation" data-cites="iIEKCTLU">(<a href="#ref-iIEKCTLU" role="doc-biblioref">Hampton et al., 2015</a>)</span>.
This strategy has been increasingly adopted in numerous studies in ecology and evolution (<em>e.g.</em>, the Zenodo repositores for <span class="citation" data-cites="GQj3c17f">Hudgins (<a href="#ref-GQj3c17f" role="doc-biblioref">2022</a>)</span>, <span class="citation" data-cites="bZNn2hbh">EllenJCoombs (<a href="#ref-bZNn2hbh" role="doc-biblioref">2020</a>)</span>, and <span class="citation" data-cites="ZI1OqZNr">Lrharper1 (<a href="#ref-ZI1OqZNr" role="doc-biblioref">2018</a>)</span>).</p>
<p>An important aspect of making code and data citable and reusable is to add an appropriate license to protect intellectual property.
Code published without a license is under exclusive copyright (by default), protecting it from copy, distribution, and modifications.
One may grant specific rights to their code for reuse by adding licensing files and specifications within GitHub repositories <span class="citation" data-cites="TOsASkn5">(<a href="#ref-TOsASkn5" role="doc-biblioref"><em>Adding a License to a Repository</em>, n.d.</a>)</span>.
The Choose a License (<a href="https://choosealicense.com/non-software/" class="uri">https://choosealicense.com/non-software/</a>) website offers further guidance on the licenses available for research and creative products.
For example, Creative Commons (CC; <a href="https://creativecommons.org/licenses/" class="uri">https://creativecommons.org/licenses/</a>) licenses can specify that shared code is intended for a specific analysis.
A CC BY 4.0 license specifies that any code (or other creative products) must be appropriately credited to its original author when distributed, adapted or reused.</p>
<h3 id="collaborative-and-asynchronous-code-editing">Collaborative and asynchronous code editing</h3>
<!--*Contributors to this section: Kaitlyn Gaynor, Rob Crystal-Ornelas, Ali, Allison Binley*-->
<p>Researchers can asynchronously communicate and individually or collaboratively work in GitHub.
By forking, branching, and cloning repositories (<a href="#definitions">Box 1</a>), researchers can simultaneously work on different issues of the same research project to then later merge changes to the main branch project with pull requests.
Explicit project organization and increased communication in GitHub Issues, Discussions, or within pull requests can help with project development and with potential merge conflicts due to users simultaneously working on the same sections <span class="citation" data-cites="1BJcvyTmV">(<a href="#ref-1BJcvyTmV" role="doc-biblioref">Vale et al., 2019</a>)</span>.
Moreover, version control <span class="citation" data-cites="K7nbP1Ty">(<a href="#ref-K7nbP1Ty" role="doc-biblioref">Crystal-Ornelas et al., 2022</a>)</span> and commit reversal features allow researchers to track and progress without worrying about irreparably modifying someone else’s work.
By enabling more comprehensive remote collaboration, GitHub encourages the exchange of ideas among researchers at different institutions and in different countries, which can serve to improve the quality of the research itself by providing open access to data and code.
In academic settings, GitHub can also facilitate interactions between research advisors and advisees, providing a platform for students or other trainees to share in-progress code, and flag specific challenges or questions for their supervisors or mentors (<a href="#tbl:roles">Table 2</a>).
Periodic code review <span class="citation" data-cites="hm9PaCLD">(<a href="#ref-hm9PaCLD" role="doc-biblioref">Song et al., 2020</a>)</span> can also help advisors identify errors early in the process, and inform further training and mentorship to fill gaps in skills.</p>
<h3 id="writing-a-manuscript">Writing a manuscript</h3>
<!--*contributors to this section:* Emma J. Hudgins-->
<p>Beyond supporting collaborative code development, GitHub can be used for writing manuscripts.
Writing a manuscript and storing its associated data and code in GitHub increases scientific reproducibility because text, code, and data can be found in one place.
Writing a manuscript on GitHub may take more time when compared to using conventional text processors <span class="citation" data-cites="4ny1onB0">(<a href="#ref-4ny1onB0" role="doc-biblioref">Ram, 2013</a>)</span>.
Nevertheless, GitHub has many features that can allow for a powerful collaborative workflow when writing manuscripts.
Text documents stored and versioned in GitHub can be instantly displayed when written in Markdown, a lightweight markup language increasingly popular among scientists.
Co-authors can contribute changes or suggest revisions to a manuscript written in GitHub through pull requests (<a href="#definitions">Box 1</a>).
Pull requests provide line-by-line views of proposed changes, which can be commented, modified, or approved by designated reviewers and collaborators (<a href="#tbl:roles">Table 2</a>).
Relevant literature or issues can be made using the Discussions and Issues features.
Moreover, real-time collaboration on text documents stored in GitHub repositories can be achieved with the help of other platforms [<em>e.g.</em>, HackMD (<a href="https://hackmd.io" class="uri">https://hackmd.io</a>) for Markdown documents].</p>
<p>We wrote this manuscript using Manubot <span class="citation" data-cites="Re6Eg2va">(<a href="#ref-Re6Eg2va" role="doc-biblioref"><em>Manubot - Manuscripts, Open and Automated</em>, n.d.</a>)</span>, a modifiable workflow implemented in GitHub to automatically render manuscripts and automate bibliographical tasks <span class="citation" data-cites="YuJbg3zO">(<a href="#ref-YuJbg3zO" role="doc-biblioref">Himmelstein et al., 2019</a>)</span>.
Manubot uses GitHub’s automation workflow, GitHub Actions, to combine and convert individual Markdown files into a single LaTeX document, which can then be converted to a Word or PDF document, and displayed as a webpage.
Citations and bibliographic references are automatically managed with citable persistent identifiers (<em>e.g.</em>, DOIs, PubMed IDs, ISBNs, URLs).
The resulting manuscript can be rendered with document templates and citation style language formatting to meet journal formatting requirements.
Every change made to the manuscript triggers its rendering, so that updates are readily displayed and made publicly available.
Additional GitHub Actions can be integrated with Manubot, such as ones creating figures or generating tables (<em>e.g.</em>, <a href="https://github.com/SORTEE-Github-Hackathon/manuscript/tree/main/.github/workflows" class="uri">https://github.com/SORTEE-Github-Hackathon/manuscript/tree/main/.github/workflows</a>).</p>
<h3 id="peer-review">Peer review</h3>
<!--*contributors to this section:* Eric R. Scott-->
<p>Peer review is the standard process for assessing and judging whether research done in ecology and evolution should be published in a scientific journal.
GitHub provides an open and transparent platform that can be used for either directly providing feedback on research products or addressing changes recommended by reviewers.
GitHub Issues can be used to organize and discuss reviewer suggestions and to assign them to co-authors (<em>e.g.</em>, <a href="https://github.com/SORTEE-Github-Hackathon/manuscript/issues?q=label%3A%22Reviewer+Comment%22+" class="uri">https://github.com/SORTEE-Github-Hackathon/manuscript/issues?q=label%3A%22Reviewer+Comment%22+</a>).
When reviewer comments are posted as separate issues, authors can comment on the issues to discuss possible changes and assign themselves to indicate which comments they intend to handle.
Co-authors can then integrate their edits and responses to reviewers using pull requests.</p>
<p>GitHub can also assist reviewers during the peer review process.
If the code associated with a manuscript is made available at the time of submission (<em>e.g.</em>, as a link to a GitHub repository within the Data Availability Statement), peer reviewers may be able to offer more comprehensive suggestions on the code and written materials, potentially recognizing errors before publication.
Certain journals or software development communities require submitted work or research code to be hosted on GitHub and their review processes make use of GitHub Issues (<em>e.g.</em>, rOpenSci (<a href="https://ropensci.org/software-review/" class="uri">https://ropensci.org/software-review/</a>), Journal of Open Source Software (<a href="https://joss.readthedocs.io/en/latest/submitting.html" class="uri">https://joss.readthedocs.io/en/latest/submitting.html</a>)).
rOpenSci’s efforts have resulted in many well-used R packages for ecology research including <code>rfishbase</code> <span class="citation" data-cites="Xsdcv6q">(<a href="#ref-Xsdcv6q" role="doc-biblioref">Boettiger et al., 2012</a>)</span> and <code>taxize</code> <span class="citation" data-cites="FVBWKkZu">(<a href="#ref-FVBWKkZu" role="doc-biblioref">Chamberlain &amp; Szöcs, 2013</a>)</span>.</p>
<!--## GitHub in EcoEvo examples (Part 3)-->
<h3 id="open-science-discussion">Open science discussion</h3>
<!--*Contributors to this section: Freddy Hillemann, Allison Binley, PHPB*-->
<p>Scientific publications often omit part of their intellectual and computational workflows, including the treatment of raw data and analytical steps (<em>e.g.</em>, model assumption testing).
Publishing data and reproducible workflows along with manuscripts can provide readers with all details about analytical steps and enable reproducing research experiments and results <span class="citation" data-cites="NOgBWVAr">(<a href="#ref-NOgBWVAr" role="doc-biblioref">Culina et al., 2020</a>)</span>.
In addition to storing data and code, GitHub repositories can provide a time-stamped (version controlled) preregistration of research plans and hypotheses.</p>
<p>Conventional research practices typically separate tasks among collaborators (<em>i.e.</em>, data entry, analysis, writing).
It is common that coauthors discuss, but do not actively verify, edit, or execute research tasks that are not their main responsibility.
GitHub can serve as a tool for open and tractable research development.
Collaborators can directly interact with code and data, inspect for errors, and potentially identify scientific misconduct prior to manuscript submission (<em>e.g.</em>, <a href="https://ecologyforthemasses.com/2020/02/04/pruittdata-and-the-ethics-of-data-in-science" class="uri">https://ecologyforthemasses.com/2020/02/04/pruittdata-and-the-ethics-of-data-in-science</a>).
Collaborators and readers are better positioned to discover erroneous or questionable findings if they have complete and transparent access to projects.</p>
<p>This transparency can be extended beyond co-authors to the entire scientific community and to the public.
Supplying code for (novel) methods that are proposed or used reduces barriers to knowledge, improving the ability of others to build on existing work.
This results in greater proliferation and accessibility for a broader audience.
Projects can make use of GitHub Discussions (<a href="https://docs.github.com/en/discussions" class="uri">https://docs.github.com/en/discussions</a>) to communicate among repository members (collaborators) and to engage with other scientists and the general public.
Moreover, researchers can also use the GitHub Community (<a href="https://github.community/" class="uri">https://github.community/</a>) forum to share expertise or request help from others on their analyses and ideas (<a href="#tbl:roles">Table 2</a>).</p>
<h3 id="project-continuity">Project continuity</h3>
<!--*Contributors to this section: BPME, VF, PHPB  -->
<p>Projects in ecology and evolution often involve graduate students, research assistants, and post-doctoral fellows, who hold limited-term positions <span class="citation" data-cites="D4C4k4ak">(<a href="#ref-D4C4k4ak" role="doc-biblioref">Fehr et al., 2021</a>)</span>.
Without clear plans on project continuity, the research code and data management upkeep tends to fall off as researchers move on to new projects or to other institutions.
Additionally, code and data can be difficult to access when kept only on personal devices <span class="citation" data-cites="19kmNxiHc">(<a href="#ref-19kmNxiHc" role="doc-biblioref">Vines et al., 2014</a>)</span>.</p>
<p>GitHub can facilitate project continuity among research code by making code handover easier <span class="citation" data-cites="D4C4k4ak 4ny1onB0">(<a href="#ref-D4C4k4ak" role="doc-biblioref">Fehr et al., 2021</a>; <a href="#ref-4ny1onB0" role="doc-biblioref">Ram, 2013</a>)</span>.
Through version control, the history of code and data from projects in ecology and evolution can be tracked accessible to future laboratory members and collaborators <span class="citation" data-cites="3DKwn1sY">(<a href="#ref-3DKwn1sY" role="doc-biblioref">Lowndes et al., 2017</a>)</span>.
Repositories and organizations can have designated data and code owners [or more appropriate, stewards; see <span class="citation" data-cites="iIEKCTLU">Hampton et al. (<a href="#ref-iIEKCTLU" role="doc-biblioref">2015</a>)</span>, <span class="citation" data-cites="s91uGRZ2"><em>About Code Owners</em> (<a href="#ref-s91uGRZ2" role="doc-biblioref">n.d.</a>)</span>], who can also can change through time allowing for the transition of code between research cohorts.
Other project collaborators can contribute to repository design and development, and their active involvement can both aid authors ability to act as guarantors of the project, and the clarity and reproducibility of the project for future users.
In (Figure <a href="#fig:github-diagram">1</a>), we highlight several elements of good repository structure, and the various ways that contributors may interact with them.</p>
<p>GitHub also allows users to describe and store information about software and code dependencies, ensuring that code can run using the same version of software as when it was initially developed.
This can be achieved through the access of GitHub repositories containing the necessary releases of such software, or through the use of automated Github Actions with specific software versions.</p>
<h3 id="automation">Automation</h3>
<!--*Contributors to this section: PHPB -->
<p>Automation has a strong potential to expand the scale and pace of research in ecology and evolution <span class="citation" data-cites="cfgXxgt1">(<a href="#ref-cfgXxgt1" role="doc-biblioref">Keitt &amp; Abelson, 2021</a>)</span>.
Automation frameworks can streamline many stages of the scientific process, including automated data collection and data validation <span class="citation" data-cites="bhDgD6lF lXvpQxeN">(<em>e.g.</em>, <a href="#ref-bhDgD6lF" role="doc-biblioref">Micheletti et al., 2021</a>; <a href="#ref-lXvpQxeN" role="doc-biblioref">Yenni et al., 2018</a>)</span>, automated data analysis <span class="citation" data-cites="Qh7xTLwz">(<em>e.g.</em>, <a href="#ref-Qh7xTLwz" role="doc-biblioref">Beaulieu-Jones &amp; Greene, 2017</a>)</span>, automated archiving and deployment of data, code and reports <span class="citation" data-cites="SirQKFIz">(<em>e.g.</em>, this manuscript, <a href="#ref-SirQKFIz" role="doc-biblioref">White et al., 2018</a>)</span>.
In this context, small modifications to code and data can be frequently committed and automatically tested, as in continuous integration and continuous deployment practices <span class="citation" data-cites="1Ep9EJL6y">(<a href="#ref-1Ep9EJL6y" role="doc-biblioref">Meyer, 2014</a>)</span>.
This allows for early detection and correction of errors, potentially improving confidence in scientific development by minimizing software errors <span class="citation" data-cites="ufw0ZdnI">(see <a href="#ref-ufw0ZdnI" role="doc-biblioref">Soergel, 2015</a>)</span>.
In addition to increasing scientific rigor and confidence in ecological software <span class="citation" data-cites="12103x16N">(<a href="#ref-12103x16N" role="doc-biblioref">Scheller et al., 2010</a>)</span>, automation can help advance more rapidly sharing ecological data and making sure the data are high quality <span class="citation" data-cites="rTbinQMj">(<a href="#ref-rTbinQMj" role="doc-biblioref">Dietze et al., 2018</a>)</span>.
Integrating automation workflows has been highly encouraged in areas of EEB, including predictive ecology <span class="citation" data-cites="mmCOSRfr">(<a href="#ref-mmCOSRfr" role="doc-biblioref">McIntire et al., 2022</a>)</span>, long-term ecological studies <span class="citation" data-cites="1CJo8lo2v RxK4CmfR">(<a href="#ref-RxK4CmfR" role="doc-biblioref">Ernest et al., 2018</a>; <a href="#ref-1CJo8lo2v" role="doc-biblioref">Yenni et al., 2019</a>)</span>, and management of species at risk information <span class="citation" data-cites="kZzfmBNu">(<a href="#ref-kZzfmBNu" role="doc-biblioref">Naujokaitis-Lewis et al., 2022</a>)</span>.</p>
<p>Automation can be integrated into GitHub repositories through the GitHub Actions feature <span class="citation" data-cites="NUXbp429">(<a href="#ref-NUXbp429" role="doc-biblioref"><em>Features • GitHub Actions</em>, n.d.</a>)</span>, or through alternative automation systems <span class="citation" data-cites="T03Api6e">(<em>e.g.</em>, Circle CI, <a href="#ref-T03Api6e" role="doc-biblioref"><em>Continuous Integration and Delivery</em>, n.d.</a>)</span>; Travis CI, <span class="citation" data-cites="YeFSbfFV"><em>Homepage | Travis CI – Start Building Today!</em> (<a href="#ref-YeFSbfFV" role="doc-biblioref">n.d.</a>)</span>].
Users can set up workflows associated with their repositories that are triggered by events (<em>e.g.</em>, push, pull request or at specified times) for remote servers to perform user-specified steps and actions.
These actions are highly configurable and have numerous applications, such as automatically running analyses and creating figures when data or code are updated, incorporating changes to websites or applications, testing modifications to software (<em>e.g.</em>, R or Python libraries).
Action workflows can be found in GitHub’s Marketplace (<a href="https://github.com/marketplace?type=actions" class="uri">https://github.com/marketplace?type=actions</a>) or, alternatively, in open user repositories.</p>
<h3 id="organizing-and-managing-teams">Organizing and managing teams</h3>
<!--*Contributors to this section: Katherine Hébert, Cole Brookson, Pedro Henrique P. Braga*-->
<p>GitHub Organizations are shared virtual spaces that allow teams to work in different repositories, while remaining tied together under a larger group, such as a laboratory, department, or project involving several teams.
Organizations allow larger projects with many steps or moving parts to be constrained to one virtual space, where outputs and sub-projects can be easily accessed and located without relying on individuals.
Because the repositories are grouped, members can reference and contribute to each others work without necessarily being part of the same repository, broadening the accessibility and longevity of code and writing contributions.</p>
<p>Contributors can be assembled into teams within an organization, which allows administrators to assign roles, tasks, and repository modification permissions to organization members.
Whereas access to repositories is usually assigned to individual contributors, Organizations facilitate the management of access permissions by allowing teams to be granted access to specific repositories.
This ensures repositories with sensitive information remain as restricted as needed, while others stay open and accessible to selected member groups.
The organization structure also allows for issue tracking and discussions related to research content and progress.</p>
<p>As an example, GitHub Organizations are particularly well-suited to host documents and projects within a laboratory, such as research compendia, codes of conduct, protocols, training documents, and other relevant documents that evolve collaboratively over time.
In this way, teams have full ownership of repositories within an organization, while ensuring that these materials stay accessible to the laboratory after people have moved on or when locally-stored data are lost.
This application extends to research centres, which may include several distinct projects that remain linked to institutions [<em>e.g.</em>, the German Centre for Integrative Biodiversity Research (iDiv, <a href="https://github.com/idiv-biodiversity" class="uri">https://github.com/idiv-biodiversity</a>)].
The team organizing the hackathon which inspired this article used a GitHub Organization (SORTEE-Github-Hackathon, <a href="https://github.com/SORTEE-Github-Hackathon" class="uri">https://github.com/SORTEE-Github-Hackathon</a>) to centralize the project development, from meeting notes to, ultimately, this manuscript.
Organizations are also convenient for hosting learning materials, including lectures or workshops, such as the Québec Centre for Biodiversity Science R Workshop Series (QCBSRworkshops, <a href="https://github.com/QCBSRworkshops" class="uri">https://github.com/QCBSRworkshops</a>) or the University of Edinburgh’s Coding Club (Coding Club, <a href="https://github.com/ourcodingclub" class="uri">https://github.com/ourcodingclub</a>), which may be continuously updated by an ever-evolving group of contributors.</p>
<h2 id="discussion">Discussion</h2>
<!--*Contributors to this section: Rob, Brandon*-->
<h3 id="the-promise-of-github-for-eeb-researchers">The promise of GitHub for EEB researchers</h3>
<p>There have been many calls for researchers outside of the software development community to adopt GitHub to improve their collaborative research <span class="citation" data-cites="1Du6fzB8g UsTxAq4f 10ghgV3S8">(<a href="#ref-UsTxAq4f" role="doc-biblioref">Anbaroğlu, 2021</a>; <a href="#ref-1Du6fzB8g" role="doc-biblioref">Crystal‐Ornelas et al., 2021</a>; <a href="#ref-10ghgV3S8" role="doc-biblioref">J. Perkel, 2016</a>)</span>.
These calls come in light of the continuous shift towards open-science and the increasing computational and data requirements in EEB.
Until now, resources and practical guidance on using GitHub within the EEB community have been dispersed in blog posts and video tutorials (<a href="#tips">Box 2</a>).
These resources have been useful for learning to use GitHub in our own work.
We expect that situating the main uses of GitHub in EEB into one medium, while adding on our personal perspectives, will be useful to the EEB community.</p>
<p>The thirteen use cases we described here can leverage GitHub to enable more transparent and collaborative research in ecology and evolution (Figure <a href="#fig:scatterblob">2</a>).
Most of these uses can be quickly integrated into the research practices of users (<em>e.g.</em>, storing data, creating virtual notebooks, making code citable).
Making stored data and code citable usually involves creating a repository on GitHub, pushing code and data, and then integrating a DOI minting service to the repository (<em>e.g.</em>, with Zenodo; see below).
On the other hand, some examples we described here, including course material development, web hosting, and automation, require greater effort and time commitment, but have the potential to make EEB research more open, accessible, and collaborative.
Managing full research projects or laboratories on GitHub require careful thought as to how to delegate tasks such as reviewing pull requests or creating issues.
For example, collaboratively authoring a paper using GitHub, as we have done here, involves a learning curve for co-authors less familiar with the intricacies of GitHub, requires overhead to set up the automation framework through GitHub Actions, and especially, the commitment from all co-authors to use GitHub when modifying and reviewing the text.
Despite the potential applications of GitHub to EEB research, we acknowledge that researchers might still look to other platforms for research collaboration.</p>
<h3 id="other-platforms-for-collaboration">Other platforms for collaboration</h3>
<!--*Contributors to this section: Rob*-->
<p>Despite its strong collaborative potential, we describe two use cases where GitHub’s features fall short of highly collaborative work emblematic of EEB research.</p>
<p>First, real-time document editing is still best performed on other platforms (<em>e.g.</em>, cloud-stored documents from Microsoft Word, Google Docs, hackMD (<a href="https://hackmd.io/" class="uri">https://hackmd.io/</a>)).
Second, operations that are dependent on other software might not be easily achievable in GitHub, such as manipulating figures or tables.
Although creating tables and figures can be done through code, users may choose other software to collaboratively brainstorm figures and tables (<em>e.g</em>., Google Slides, Google Sheets; but see GitHub Discussions).</p>
<h3 id="why-arent-more-eeb-researchers-using-github">Why aren’t more EEB researchers using GitHub?</h3>
<!--*Contributors to this section: Saeed, Vivienne, PHPB*-->
<p>Although GitHub has been available as a platform for more than a decade, its uptake among EEB researchers, especially as a tool for collaboration, has been slow.</p>
<p>Here, we discuss four potential barriers to GitHub use in EEB.</p>
<p>First, there may be hesitation to independently adopting and learning a new tool.
Institutional encouragement and instructional resources focused on researchers in ecology and evolution may be limited.
When GitHub is taught within an EEB context, it usually accompanies coursework in topics such as statistical programming.
It can be challenging to learn Git alongside scripting languages, statistical theory, and file system navigation, especially when many may be inexperienced with programming.
Instructors likewise may confuse the expected digital literacy of students with computational fluency, even when modern technology increasingly abstracts concepts through search optimization and preponderant integrated development environments (IDE), or ‘point-and-click’ user interfaces.</p>
<p>Second, while EEB researchers individually use GitHub, collaborative use may lag due to researchers traditionally dividing labor within projects.
Despite broad utility, GitHub remains a tool predominantly used by computer scientists and software developers.
EEB researchers may take the view that GitHub is a platform that only needs to be used by individuals writing code and may silo those aspects of projects to a single individual.
Those assumptions may obscure the utility of GitHub for tasks other than traditional data analysis and code development.
However, we emphasize that there are opportunities for collaboration using GitHub by researchers of all skill levels or time constraints (<a href="#tbl:roles">Table 2</a>); for example, project stakeholders can provide a list of use-cases or highlight important conceptual components of a project using GitHub Issues or Discussions features.</p>
<p>A third barrier may come from general reluctance to share data and code publicly, or technical and logistical issues <span class="citation" data-cites="pq2Tv1BC">(<a href="#ref-pq2Tv1BC" role="doc-biblioref">Gomes et al., 2022</a>)</span>.
GitHub is, by default, a public and open platform.
This openness may add additional pressure to students and scientists learning to use the platform.
File storage is size-limited on GitHub (but note that large files stored locally can still be versioned; see section “Storing and sharing research compendia”), requiring the complementation of other tools to fully integration project files and GitHub repositories (<em>e.g.</em>, <span class="citation" data-cites="lx49NGto">(<a href="#ref-lx49NGto" role="doc-biblioref"><em>Connect GitHub to a Project - OSF Support</em>, n.d.</a>)</span>).
A major additional barrier to EEB researchers is the lack language-specific resources for non-English speaking researchers working in ecology and evolution.
Language is a well-known obstacle to international collaborative research progress and to widespread scientific knowledge <span class="citation" data-cites="cW7vGddM">(see <a href="#ref-cW7vGddM" role="doc-biblioref">Khelifa et al., 2022</a>)</span>.
Non-English speaking EEB researchers can potentially miss opportunities to fully integrate version control, reproducibility, and other benefits of GitHub without language-inclusive contents.</p>
<p>Lastly, when projects become highly collaborative they may have to pay for GitHub features such as branch protections, multiple reviewers of pull requests, and time in its automation tools.
Currently, GitHub offers Education Packs (<a href="https://education.github.com/" class="uri">https://education.github.com/</a>) to students and academics, which extends some paid features to the free plan.
However, the acquisition of GitHub by Microsoft has raised concerns over the future of free plans, causing several biodiversity data managers to shift to alternative Open Source Git services (<em>e.g.</em>, Bitbucket and GitLab).</p>
<h3 id="tips">Box 2</h3>
<!--*Contributors to this section: Ali, Emma*-->
<table>
<colgroup>
<col style="width: 100%" />
</colgroup>
<thead>
<tr class="header">
<th>Ten tips for getting started in GitHub</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1. <strong>Check for existing solutions to your problem.</strong> The GitHub Help webpage (<a href="https://docs.github.com/en" class="uri">https://docs.github.com/en</a>) contains extensive and detailed documents with helpful screenshots. It is a good starting point for handling an issue, and has troubleshooting tips for specific problems. Alternatively, consider Tweeting your issue. There is a large community of GitHub users around the world who have likely faced analogous problems and may be able to provide quick solutions. Third, try to follow blogs (<em>e.g.</em>, <a href="https://github.blog" class="uri">https://github.blog</a>), Twitter accounts or YouTube channels that regularly post practical solutions about the most widely-used web platform for common GitHub issues.</td>
</tr>
<tr class="even">
<td>2. <strong>Consider taking free courses</strong>, such as those from Software Carpentry <span class="citation" data-cites="pjy75gHr">(<a href="#ref-pjy75gHr" role="doc-biblioref">Madicken Munk et al., 2019</a>)</span>, and sharing these courses with your lab members or colleagues.</td>
</tr>
<tr class="odd">
<td>3. <strong>Take advantage of GitHub as an asynchronous working tool for team-based projects.</strong> See the repository for this paper (<a href="https://github.com/SORTEE-Github-Hackathon/manuscript" class="uri">https://github.com/SORTEE-Github-Hackathon/manuscript</a>) as an example of a collaboratively authored manuscript that used the GitHub Discussions, Issues, Pages, and Actions features.</td>
</tr>
<tr class="even">
<td>4. <strong>Use the interactive courses from the GitHub Skills page (<a href="https://skills.github.com/" class="uri">https://skills.github.com/</a>)</strong>, which allow you to learn GitHub basics through short projects and tasks with step-by-step guides.</td>
</tr>
<tr class="odd">
<td>5. <strong>Learn markdown and use cheatsheets (<em>e.g.</em>, <a href="http://markdownguide.org/basic-syntax" class="uri">http://markdownguide.org/basic-syntax</a>)</strong> so you can write clear metadata README files for your repositories.</td>
</tr>
<tr class="even">
<td>6. <strong>Consult online resources.</strong> The Jenny Bryan Universe of GitHub material, for example, provides a thorough and accessible introduction for a multitude of research-related uses for GitHub, and includes a book <span class="citation" data-cites="ZvrOcg9w">(<a href="#ref-ZvrOcg9w" role="doc-biblioref">Hester, n.d.</a>)</span>, statistics course <span class="citation" data-cites="6CMMeSeD">(<a href="#ref-6CMMeSeD" role="doc-biblioref">Bryan &amp; TAs, n.d.</a>)</span> and academic article <span class="citation" data-cites="RVetqmsg">(<a href="#ref-RVetqmsg" role="doc-biblioref">Bryan, 2018</a>)</span>.</td>
</tr>
<tr class="odd">
<td>7. <strong>Do not be afraid of trial-and-error.</strong> One of the best ways to learn GitHub is the “trial-and-error” method. Learning from your own mistakes can be a valuable way to master your GitHub abilities. In any case, if you make mistakes, GitHub allows you to revert any steps that you desire via version controlling.</td>
</tr>
<tr class="even">
<td>8. <strong>If you are an educator, include lectures on reproducibility and tools for creating reproducible workflows in the curricula.</strong> Some graduate programs include coursework on course R Markdown and GitHub. Getting students started with these tools earlier will prevent the resistance that comes from working with a less reproducible workflow for a longer period of time (see example <a href="https://github.com/rmcelreath/stat_rethinking_2022" class="uri">https://github.com/rmcelreath/stat_rethinking_2022</a>).</td>
</tr>
<tr class="odd">
<td>9. <strong>Try to begin committing with graphical user interfaces (GUI) instead of command line interfaces (CLI).</strong> Examples of GUI are the GitHub Desktop (<a href="https://desktop.github.com" class="uri">https://desktop.github.com</a>), git-gui (<a href="https://git-scm.com/docs/git-gui" class="uri">https://git-scm.com/docs/git-gui</a>), RStudio (<a href="https://www.rstudio.com" class="uri">https://www.rstudio.com</a>), Visual Studio Code (<a href="https://code.visualstudio.com" class="uri">https://code.visualstudio.com</a>), Atom (<a href="https://atom.io" class="uri">https://atom.io</a>), GitKraken (<a href="https://www.gitkraken.com" class="uri">https://www.gitkraken.com</a>).</td>
</tr>
<tr class="even">
<td>10. <strong>Get help deciphering GitHub Notifications.</strong> Try using tools like Octobox (<a href="https://octobox.io" class="uri">https://octobox.io</a>) to disentangle and manage multiple notifications from distinct GitHub projects.</td>
</tr>
</tbody>
</table>
<h2 id="conclusion">Conclusion</h2>
<p>We describe thirteen practical ways that ecologists and evolutionary biologists can adopt GitHub to improve their research workflow and make it more open and reproducible.
We provide definitions (<a href="#definitions">Box 1</a>) and types of users (Figure <a href="#fig:github-diagram">1</a>) to help researchers identify and prioritize the skills and tools to learn and apply.
We highlight tools providing high collaborative potential (<em>e.g.</em>, open science discussion, collaborative code editing) to more individual focused (<em>e.g.</em>, storing code and data, building a website).
We argue that the tools readily available in GitHub have the potential to make ecology and evolution more open, reproducible and transparent.
With this comprehensive review of how EEB researchers can use GitHub, we encourage researchers at any career stage to adopt GitHub as a platform for sharing and collaboration.</p>
<h2 id="author-contributions">Author Contributions</h2>
<p>We indicate author contributions using the <a href="https://casrai.org/credit/">CRediT Taxonomy</a>.</p>
<ul>
<li>Conceptualization: RCO, BPME, KH, EJH, LLSR, PHPB</li>
<li>Investigation: RCO, BPME, KH, EJH, LLSR, ERS, MJG, VF, ADB, CBB, KMG, SSHS, AG, FH, HW, DGEG, PHPB</li>
<li>Methodology: RCO, BPME, KH, EJH, LLSR, PHPB</li>
<li>Project administration: RCO</li>
<li>Software: RCO, BPME, KH, EJH, LLSR, ERS, MJG, VF, ADB, CBB, KMG, SSHS, AG, FH, HW, DGEG, PHPB</li>
<li>Visualization: RCO, BPME, KH, EJH, LLSR, ERS, MJG, VF, ADB, CBB, KMG, SSHS, AG, FH, HW, DGEG, PHPB</li>
<li>Writing – original draft: RCO, BPME, KH, EJH, LLSR, ERS, MJG, VF, ADB, CBB, KMG, SSHS, AG, FH, HW, DGEG, PHPB</li>
<li>Writing – review and editing: RCO, BPME, KH, EJH, LLSR, ERS, MJG, VF, ADB, CBB, KMG, SSHS, AG, FH, HW, DGEG, PHPB</li>
<li>Resources: PHPB</li>
</ul>
<h2 id="acknowledgements">Acknowledgements</h2>
<p>This manuscript arose from a hackathon at the Society for Open, Reliable, and Transparent Ecology and Evolution (SORTEE) virtual meeting in July 2021.
We thank Ciera Martinez for being a co-organizer of the SORTEE hackathon that started our discussion on GitHub in EEB and for creating our hackathon website on GitHub.</p>
<p>RCO was funded by the U.S. Department of Energy, Office of Science, Office of Biological and Environmental Research, Earth and Environmental Sciences Division, Data Management program under contract number DE-AC02-05CH11231.</p>
<h2 id="code-and-data-availability">Code and data availability</h2>
<p>The source code and data for this manuscript are available at <a href="https://github.com/SORTEE-Github-Hackathon/manuscript" class="uri">https://github.com/SORTEE-Github-Hackathon/manuscript</a>.
The source code will be archived in the ESS-DIVE data repository prior to publication.</p>
<h2 id="ethics-declarations---competing-interests">Ethics declarations - Competing interests</h2>
<p>RCO was recently offered a position to work at GitHub (on July 15, 2022) and will start employment at GitHub on August 23, 2022.
Initial discussion of publishing this manuscript began in July 2021 and all work on the manuscript took place while RCO was an employee at Lawrence Berkeley National Laboratory.</p>
<h2 class="page_break_before" id="figures">Figures</h2>
<div id="fig:github-diagram" class="fignos">
<figure>
<img src="images/github_diagram.png" alt="Figure 1: An overview of Git’s core features. A) Multi-faceted components allow for code writing, small data storage, manuscript writing, and project management to all be done in one place. CONTRIBUTING.md, LICENCE.md, and README.md files allow new team members, or others wanting to use materials, to understand the project components and learn how they can engage with the project and existing team members. B) Issues, Pull Requests, Discussions, and Projects allow for team members to ask for feedback, suggest fixes, discuss related ideas, and keep track of all the moving parts of a project. C) All collaborators on a project can be a part of a single repository, with varying push privileges and responsibilities" />
<figcaption aria-hidden="true"><span>Figure 1:</span> An overview of Git’s core features. A) Multi-faceted components allow for code writing, small data storage, manuscript writing, and project management to all be done in one place. <code>CONTRIBUTING.md</code>, <code>LICENCE.md</code>, and <code>README.md</code> files allow new team members, or others wanting to use materials, to understand the project components and learn how they can engage with the project and existing team members. B) Issues, Pull Requests, Discussions, and Projects allow for team members to ask for feedback, suggest fixes, discuss related ideas, and keep track of all the moving parts of a project. C) All collaborators on a project can be a part of a single repository, with varying push privileges and responsibilities</figcaption>
</figure>
</div>
<!--Reference this figure in text like this (Figure @fig:github-diagram).-->
<div id="fig:scatterblob" class="fignos">
<figure>
<img src="images/scatterblob_1-viridis-turbo.png" alt="Figure 2: A summary of ways GitHub can be used showing technical difficulty and degree of collaboration for each. Activities higher on the vertical axis require usage knowledge of more GitHub features than activities lower on the axis. On the horizontal axis, each activity spans a region representing who is potentially involved with or benefits from each activity. For example, storing data and code mainly benefits individual researchers or members of a lab group while making data and code citable and reproducible benefit other labs and the larger community as well. Independently of a users knowledge level of GitHub features, there are ways to use GitHub that allow tapping unto one of the most salient benefits of the platform: facilitating and enhancing collaboration." />
<figcaption aria-hidden="true"><span>Figure 2:</span> A summary of ways GitHub can be used showing technical difficulty and degree of collaboration for each. Activities higher on the vertical axis require usage knowledge of more GitHub features than activities lower on the axis. On the horizontal axis, each activity spans a region representing who is potentially involved with or benefits from each activity. For example, storing data and code mainly benefits individual researchers or members of a lab group while making data and code citable and reproducible benefit other labs and the larger community as well. Independently of a users knowledge level of GitHub features, there are ways to use GitHub that allow tapping unto one of the most salient benefits of the platform: facilitating and enhancing collaboration.</figcaption>
</figure>
</div>
<!--Reference this figure in text like this (Figure @fig:scatterblob).-->
<h2 id="tables">Tables</h2>
<!--DO NOT EDIT. Generated from R/comparison_table.Rmd, so edit that and render it to make changes.  This should get tacked on to the end of 09.figures-tables.md when compiled by Manubot -->
<div id="tbl:compare" class="tablenos">
<table id="tbl:compare">
<caption><span>Table 1:</span> A comparison of technologies commonly used for collaborating on research in Ecology and Evolutionary Biology. In the first column, we group platforms for collaboration into broad guilds. The second column lists the platform for collaboration. The remaining columns indicate whether the platform for collaboration includes certain features. </caption>
<colgroup>
<col style="width: 6%" />
<col style="width: 5%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 3%" />
<col style="width: 5%" />
<col style="width: 17%" />
<col style="width: 16%" />
<col style="width: 25%" />
<col style="width: 13%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Guild</th>
<th style="text-align: left;">Software</th>
<th style="text-align: left;">Version control</th>
<th style="text-align: left;">Backup (cloud)</th>
<th style="text-align: left;">Passive collaboration</th>
<th style="text-align: left;">Active real-time collaboration</th>
<th style="text-align: left;">Free plan available</th>
<th style="text-align: left;">Permanent (DOI)</th>
<th style="text-align: left;">Storage limits</th>
<th style="text-align: left;">GitHub Integration</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Multi-tool</td>
<td style="text-align: left;">GitHub</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">no</td>
<td style="text-align: left;">Broadly used free version. Advanced features are provided for free to students and education professionals.</td>
<td style="text-align: left;">A DOI can only be obtained when integrating to other services that can mint DOI (e.g., Zenodo, OSF).</td>
<td style="text-align: left;">100MB per file, 500MB per private repository (2GB for paid accounts). 100GB for public repositories. Larger files (up to 2GB) can be attached to releases</td>
<td style="text-align: left;">N/A</td>
</tr>
<tr class="even">
<td style="text-align: left;">Multi-tool</td>
<td style="text-align: left;">Open Science Framework</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">25GB for private projects, up to 5GB per file, plus partner add-ons, 50GB for public projects</td>
<td style="text-align: left;">yes</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Long-term (public) data repositories</td>
<td style="text-align: left;">PANGAEA</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">no</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">10 GB free</td>
<td style="text-align: left;">no</td>
</tr>
<tr class="even">
<td style="text-align: left;">Long-term (public) data repositories</td>
<td style="text-align: left;">Zenodo</td>
<td style="text-align: left;">after published</td>
<td style="text-align: left;">after published</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">no</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">50 GB per dataset</td>
<td style="text-align: left;">yes</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Long-term (public) data repositories</td>
<td style="text-align: left;">Dryad</td>
<td style="text-align: left;">after published</td>
<td style="text-align: left;">after published</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">no</td>
<td style="text-align: left;">some journals cover cost</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">300 GB per publication</td>
<td style="text-align: left;">Can link to individual files (not entire repository), thus not fully integrated</td>
</tr>
<tr class="even">
<td style="text-align: left;">Long-term (public) data repositories</td>
<td style="text-align: left;">Figshare</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">no</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">20 GB free, up to 5 TB</td>
<td style="text-align: left;">yes</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Temporary (personal) drive storage</td>
<td style="text-align: left;">Google Drive</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">limited free version &amp; paid</td>
<td style="text-align: left;">no</td>
<td style="text-align: left;">15GB free, up to 100GB with Google One</td>
<td style="text-align: left;">yes</td>
</tr>
<tr class="even">
<td style="text-align: left;">Temporary (personal) drive storage</td>
<td style="text-align: left;">Box</td>
<td style="text-align: left;">limited</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">no</td>
<td style="text-align: left;">no</td>
<td style="text-align: left;">Unlimited total size for subscription</td>
<td style="text-align: left;">yes</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Temporary (personal) drive storage</td>
<td style="text-align: left;">DropBox</td>
<td style="text-align: left;">limited</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">limited free version &amp; paid</td>
<td style="text-align: left;">no</td>
<td style="text-align: left;">2GB free</td>
<td style="text-align: left;">yes</td>
</tr>
<tr class="even">
<td style="text-align: left;">Temporary (personal) drive storage</td>
<td style="text-align: left;">OneDrive and the Office Suite</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">limited free version &amp; paid</td>
<td style="text-align: left;">no</td>
<td style="text-align: left;">5 GB free, up to 1TB paid</td>
<td style="text-align: left;">yes</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Collaborative code/text editors</td>
<td style="text-align: left;">Overleaf (online latex editor)</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">no</td>
<td style="text-align: left;">1MB for individual .tex, 50MB for individual files, unlimited project size</td>
<td style="text-align: left;">yes</td>
</tr>
<tr class="even">
<td style="text-align: left;">Collaborative code/text editors</td>
<td style="text-align: left;">Jupyter Notebook</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">?</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">with Colab</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">no</td>
<td style="text-align: left;">via Binder: no hard limit, but suggests no files &gt;100MB, can also store on GitHub or Google Colab</td>
<td style="text-align: left;">yes</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Collaborative code/text editors</td>
<td style="text-align: left;">HackMD</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">yes</td>
<td style="text-align: left;">limited free version &amp; paid</td>
<td style="text-align: left;">no</td>
<td style="text-align: left;">3 documents free, private invitee limits</td>
<td style="text-align: left;">yes</td>
</tr>
</tbody>
</table>
</div>
<!-- An example citation of this table (Table @tbl:compare). -->
<!--DO NOT EDIT. Generated from R/roles_table.Rmd, so edit that and render it to make changes.  This should get tacked on to the end of 09.figures-tables.md when compiled by Manubot -->
<div id="tbl:roles" class="tablenos">
<table id="tbl:roles">
<caption><span>Table 2:</span> A non-exhaustive collection of ideas for how various GitHub
features could be utilized for a research project. Here we have
categorized contributors/collaborators into five roles. A Project
Manager owns the GitHub repository for a project, and leads the academic
project (e.g., lead author of a manuscript). A co-author contributes to
writing and other aspects of research, but may have limited or no
experience with programming, git, and/or GitHub. A code contributor
writes or edits analysis code for the project. A code reviewer could be
a project collaborator or a peer reviewer who reviews project code. They
are familiar with coding, but not necessarily with git or GitHub (but
they are willing to learn). Finally, community members could be other
researchers or non-researchers interested in reproducing results,
re-using code or data, or communicating with researchers involved in the
project. These roles are not mutually exclusive—a co-author could also
be e code contributor and code reviewer, for example. For definitions of
the GitHub features, see Box 1. </caption>
<colgroup>
<col style="width: 4%" />
<col style="width: 14%" />
<col style="width: 10%" />
<col style="width: 15%" />
<col style="width: 10%" />
<col style="width: 14%" />
<col style="width: 13%" />
<col style="width: 15%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Role</th>
<th style="text-align: left;">GitHub repository</th>
<th style="text-align: left;">README</th>
<th style="text-align: left;">Issue</th>
<th style="text-align: left;">Discussion</th>
<th style="text-align: left;">Pull Request</th>
<th style="text-align: left;">Fork</th>
<th style="text-align: left;">GitHub Pages</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Project manager</td>
<td style="text-align: left;">Set contributor permissions, share code of conduct</td>
<td style="text-align: left;">Project description, citation, DOIs</td>
<td style="text-align: left;">Assign tasks to collaborators</td>
<td style="text-align: left;">Discuss project directions and goals</td>
<td style="text-align: left;">Approve and incorporate edits to code and/or writing</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">Share up-to-date reports, figures, or draft manuscript</td>
</tr>
<tr class="even">
<td style="text-align: left;">Co-author</td>
<td style="text-align: left;">Edit Markdown text or add files</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">Propose changes involving code (e.g. analyses, figures)</td>
<td style="text-align: left;">Discuss proposed changes to manuscript</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Code contributor</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">Suggest code changes</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">Contribute changes to code, initiate code review</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">Contribute to project website</td>
</tr>
<tr class="even">
<td style="text-align: left;">Code reviewer</td>
<td style="text-align: left;">Find all code related to a project</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">Highlight specific lines of code and make suggestions</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">Review or recommended changes in code</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Community</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">Suggest additional features and report bugs</td>
<td style="text-align: left;">Ask questions about data and code</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">Create a linked, editable copy of the repository</td>
<td style="text-align: left;">View project website</td>
</tr>
</tbody>
</table>
</div>
<!-- An example citation of this table (Table @tbl:roles). -->
<h2 class="page_break_before" id="references">References</h2>
<!-- Explicitly insert bibliography here -->
<div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="doc-bibliography">
<div id="ref-s91uGRZ2" class="csl-entry" role="doc-biblioentry">
<em>About code owners</em>. (n.d.). GitHub Docs. Retrieved November 17, 2022, from <a href="https://ghdocs-prod.azurewebsites.net/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners">https://ghdocs-prod.azurewebsites.net/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners</a>
</div>
<div id="ref-1Co6ZZjF1" class="csl-entry" role="doc-biblioentry">
<em>About large files on GitHub</em>. (n.d.). GitHub Docs. Retrieved November 17, 2022, from <a href="https://ghdocs-prod.azurewebsites.net/en/repositories/working-with-files/managing-large-files/about-large-files-on-github">https://ghdocs-prod.azurewebsites.net/en/repositories/working-with-files/managing-large-files/about-large-files-on-github</a>
</div>
<div id="ref-RhBKe0MG" class="csl-entry" role="doc-biblioentry">
<em>About Projects</em>. (n.d.). GitHub Docs. Retrieved November 17, 2022, from <a href="https://ghdocs-prod.azurewebsites.net/en/issues/planning-and-tracking-with-projects/learning-about-projects/about-projects">https://ghdocs-prod.azurewebsites.net/en/issues/planning-and-tracking-with-projects/learning-about-projects/about-projects</a>
</div>
<div id="ref-TOsASkn5" class="csl-entry" role="doc-biblioentry">
<em>Adding a license to a repository</em>. (n.d.). GitHub Docs. Retrieved November 17, 2022, from <a href="https://ghdocs-prod.azurewebsites.net/en/communities/setting-up-your-project-for-healthy-contributions/adding-a-license-to-a-repository">https://ghdocs-prod.azurewebsites.net/en/communities/setting-up-your-project-for-healthy-contributions/adding-a-license-to-a-repository</a>
</div>
<div id="ref-13QX8XU3J" class="csl-entry" role="doc-biblioentry">
Alston, J. M., &amp; Rick, J. A. (2021). A Beginner's Guide to Conducting Reproducible Research. <em>The Bulletin of the Ecological Society of America</em>, <em>102</em>(2). <a href="https://doi.org/10.1002/bes2.1801">https://doi.org/10.1002/bes2.1801</a>
</div>
<div id="ref-UsTxAq4f" class="csl-entry" role="doc-biblioentry">
Anbaroğlu, B. (2021). A collaborative GIS programming course using GitHub Classroom. <em>Transactions in GIS</em>, <em>25</em>(6), 3132–3158. <a href="https://doi.org/10.1111/tgis.12810">https://doi.org/10.1111/tgis.12810</a>
</div>
<div id="ref-1HZdsK5Kn" class="csl-entry" role="doc-biblioentry">
Baker, M. (2016). 1,500 scientists lift the lid on reproducibility. <em>Nature</em>, <em>533</em>(7604), 452–454. <a href="https://doi.org/10.1038/533452a">https://doi.org/10.1038/533452a</a>
</div>
<div id="ref-Qh7xTLwz" class="csl-entry" role="doc-biblioentry">
Beaulieu-Jones, B. K., &amp; Greene, C. S. (2017). Reproducibility of computational workflows is automated using continuous analysis. <em>Nature Biotechnology</em>, <em>35</em>(4), 342–346. <a href="https://doi.org/10.1038/nbt.3780">https://doi.org/10.1038/nbt.3780</a>
</div>
<div id="ref-Xsdcv6q" class="csl-entry" role="doc-biblioentry">
Boettiger, C., Lang, D. T., &amp; Wainwright, P. C. (2012). rfishbase: exploring, manipulating and visualizing FishBase data from R. <em>Journal of Fish Biology</em>, <em>81</em>(6), 2030–2039. <a href="https://doi.org/10.1111/j.1095-8649.2012.03464.x">https://doi.org/10.1111/j.1095-8649.2012.03464.x</a>
</div>
<div id="ref-ydrk01SR" class="csl-entry" role="doc-biblioentry">
Briney, K., Coates, H., &amp; Goben, A. (2020). Foundational Practices of Research Data Management. <em>Research Ideas and Outcomes</em>, <em>6</em>. <a href="https://doi.org/10.3897/rio.6.e56508">https://doi.org/10.3897/rio.6.e56508</a>
</div>
<div id="ref-RVetqmsg" class="csl-entry" role="doc-biblioentry">
Bryan, J. (2018). Excuse Me, Do You Have a Moment to Talk About Version Control? <em>The American Statistician</em>, <em>72</em>(1), 20–27. <a href="https://doi.org/10.1080/00031305.2017.1399928">https://doi.org/10.1080/00031305.2017.1399928</a>
</div>
<div id="ref-6CMMeSeD" class="csl-entry" role="doc-biblioentry">
Bryan, J., &amp; TAs, T. S. 545. (n.d.). <em>STAT 545</em>. Retrieved November 17, 2022, from <a href="https://stat545.com/">https://stat545.com/</a>
</div>
<div id="ref-nwCtHDCn" class="csl-entry" role="doc-biblioentry">
<em>Build software better, together</em>. (n.d.). GitHub. Retrieved November 17, 2022, from <a href="https://github.com">https://github.com</a>
</div>
<div id="ref-FVBWKkZu" class="csl-entry" role="doc-biblioentry">
Chamberlain, S. A., &amp; Szöcs, E. (2013). taxize: taxonomic search and retrieval in R. <em>F1000Research</em>, <em>2</em>, 191. <a href="https://doi.org/10.12688/f1000research.2-191.v2">https://doi.org/10.12688/f1000research.2-191.v2</a>
</div>
<div id="ref-lx49NGto" class="csl-entry" role="doc-biblioentry">
<em>Connect GitHub to a Project - OSF Support</em>. (n.d.). Retrieved November 17, 2022, from <a href="https://help.osf.io/article/211-connect-github-to-a-project">https://help.osf.io/article/211-connect-github-to-a-project</a>
</div>
<div id="ref-T03Api6e" class="csl-entry" role="doc-biblioentry">
<em>Continuous Integration and Delivery</em>. (n.d.). CircleCI. Retrieved November 17, 2022, from <a href="https://circleci.com/">https://circleci.com/</a>
</div>
<div id="ref-K7nbP1Ty" class="csl-entry" role="doc-biblioentry">
Crystal-Ornelas, R., Edwards, B. P. M., Hébert, K., Hudgins, E. J., Reyes, L. L. S., Scott, E. R., Grainger, M. J., Foroughirad, V., Binley, A. D., Brookson, C. B., Gaynor, K. M., Sabet, S. S., Güncan, A., Hillemann, F., Weierbach, H., Gomes, D. G. E., &amp; Braga, P. H. P. (2022). <em>Not just for programmers: How GitHub can accelerate collaborative and reproducible research in ecology and evolution</em>. Manubot. <a href="https://SORTEE-Github-Hackathon.github.io/manuscript/">https://SORTEE-Github-Hackathon.github.io/manuscript/</a>
</div>
<div id="ref-1Du6fzB8g" class="csl-entry" role="doc-biblioentry">
Crystal‐Ornelas, R., Varadharajan, C., Bond‐Lamberty, B., Boye, K., Burrus, M., Cholia, S., Crow, M., Damerow, J., Devarakonda, R., Ely, K. S., Goldman, A., Heinz, S., Hendrix, V., Kakalia, Z., Pennington, S. C., Robles, E., Rogers, A., Simmonds, M., Velliquette, T., … Agarwal, D. A. (2021). A Guide to Using GitHub for Developing and Versioning Data Standards and Reporting Formats. <em>Earth and Space Science</em>, <em>8</em>(8). <a href="https://doi.org/10.1029/2021ea001797">https://doi.org/10.1029/2021ea001797</a>
</div>
<div id="ref-NOgBWVAr" class="csl-entry" role="doc-biblioentry">
Culina, A., van den Berg, I., Evans, S., &amp; Sánchez-Tójar, A. (2020). Low availability of code in ecology: A call for urgent action. <em>PLOS Biology</em>, <em>18</em>(7), e3000763. <a href="https://doi.org/10.1371/journal.pbio.3000763">https://doi.org/10.1371/journal.pbio.3000763</a>
</div>
<div id="ref-MXxgZJ45" class="csl-entry" role="doc-biblioentry">
Dawson, C. (2016). <em>Building tools with GitHub: customize your workflow</em> (First Edition). O'Reilly.
</div>
<div id="ref-rTbinQMj" class="csl-entry" role="doc-biblioentry">
Dietze, M. C., Fox, A., Beck-Johnson, L. M., Betancourt, J. L., Hooten, M. B., Jarnevich, C. S., Keitt, T. H., Kenney, M. A., Laney, C. M., Larsen, L. G., Loescher, H. W., Lunch, C. K., Pijanowski, B. C., Randerson, J. T., Read, E. K., Tredennick, A. T., Vargas, R., Weathers, K. C., &amp; White, E. P. (2018). Iterative near-term ecological forecasting: Needs, opportunities, and challenges. <em>Proceedings of the National Academy of Sciences</em>, <em>115</em>(7), 1424–1432. <a href="https://doi.org/10.1073/pnas.1710231115">https://doi.org/10.1073/pnas.1710231115</a>
</div>
<div id="ref-bZNn2hbh" class="csl-entry" role="doc-biblioentry">
EllenJCoombs. (2020). <em>EllenJCoombs/Asymmetry-evolution-cetaceans-for-publication-v.1.0.0</em> (Version v.1.0.0) [Computer software]. Zenodo. <a href="https://doi.org/10.5281/zenodo.3893943">https://doi.org/10.5281/zenodo.3893943</a>
</div>
<div id="ref-RxK4CmfR" class="csl-entry" role="doc-biblioentry">
Ernest, S. K. M., Yenni, G. M., Allington, G., Bledsoe, E. K., Christensen, E. M., Diaz, R. M., Geluso, K., Goheen, J. R., Guo, Q., Heske, E., Kelt, D., Meiners, J. M., Munger, J., Restrepo, C., Samson, D. A., Schutzenhofer, M. R., Skupski, M., Supp, S. R., Thibault, K., … Valone, T. J. (2018). <em>The Portal Project: a long-term study of a Chihuahuan desert ecosystem</em>. Cold Spring Harbor Laboratory. <a href="https://doi.org/10.1101/332783">https://doi.org/10.1101/332783</a>
</div>
<div id="ref-NUXbp429" class="csl-entry" role="doc-biblioentry">
<em>Features • GitHub Actions</em>. (n.d.). GitHub. Retrieved November 17, 2022, from <a href="https://github.com/features/actions">https://github.com/features/actions</a>
</div>
<div id="ref-D4C4k4ak" class="csl-entry" role="doc-biblioentry">
Fehr, J., Himpe, C., Rave, S., &amp; Saak, J. (2021). Sustainable Research Software Hand-Over. <em>Journal of Open Research Software</em>, <em>9</em>(1), 5. <a href="https://doi.org/10.5334/jors.307">https://doi.org/10.5334/jors.307</a>
</div>
<div id="ref-SLq38RVv" class="csl-entry" role="doc-biblioentry">
Figueiredo, A. S. (2017). Data Sharing: Convert Challenges into Opportunities. <em>Frontiers in Public Health</em>, <em>5</em>. <a href="https://doi.org/10.3389/fpubh.2017.00327">https://doi.org/10.3389/fpubh.2017.00327</a>
</div>
<div id="ref-pq2Tv1BC" class="csl-entry" role="doc-biblioentry">
Gomes, D. G. E., Pottier, P., Crystal-Ornelas, R., Hudgins, E. J., Foroughirad, V., Sánchez-Reyes, L. L., Turba, R., Martinez, P. A., Moreau, D., Bertram, M., Smout, C., &amp; Gaynor, K. (2022). <em>Why don’t we share data and code? Perceived barriers and benefits to public archiving practices</em>. Center for Open Science. <a href="https://doi.org/10.31222/osf.io/gaj43">https://doi.org/10.31222/osf.io/gaj43</a>
</div>
<div id="ref-1HhzKAC1K" class="csl-entry" role="doc-biblioentry">
Goring, S. J., Weathers, K. C., Dodds, W. K., Soranno, P. A., Sweet, L. C., Cheruvelil, K. S., Kominoski, J. S., Rüegg, J., Thorn, A. M., &amp; Utz, R. M. (2014). Improving the culture of interdisciplinary collaboration in ecology by expanding measures of success. <em>Frontiers in Ecology and the Environment</em>, <em>12</em>(1), 39–47. <a href="https://doi.org/10.1890/120370">https://doi.org/10.1890/120370</a>
</div>
<div id="ref-iIEKCTLU" class="csl-entry" role="doc-biblioentry">
Hampton, S. E., Anderson, S. S., Bagby, S. C., Gries, C., Han, X., Hart, E. M., Jones, M. B., Lenhardt, W. C., MacDonald, A., Michener, W. K., Mudge, J., Pourmokhtarian, A., Schildhauer, M. P., Woo, K. H., &amp; Zimmerman, N. (2015). The Tao of open science for ecology. <em>Ecosphere</em>, <em>6</em>(7), art120. <a href="https://doi.org/10.1890/es14-00402.1">https://doi.org/10.1890/es14-00402.1</a>
</div>
<div id="ref-fJWFe93e" class="csl-entry" role="doc-biblioentry">
Hannay, J. E., MacLeod, C., Singer, J., Langtangen, H. P., Pfahl, D., &amp; Wilson, G. (2009, May). How do scientists develop and use scientific software? <em>2009 ICSE Workshop on Software Engineering for Computational Science and Engineering</em>. 2009 ICSE Workshop on Software Engineering for Computational Science and Engineering (SECSE). <a href="https://doi.org/10.1109/secse.2009.5069155">https://doi.org/10.1109/secse.2009.5069155</a>
</div>
<div id="ref-ZvrOcg9w" class="csl-entry" role="doc-biblioentry">
Hester, J. B., the STAT 545 TAs, Jim. (n.d.). <em>Let’s Git started | Happy Git and GitHub for the useR</em>. Retrieved November 17, 2022, from <a href="https://happygitwithr.com/">https://happygitwithr.com/</a>
</div>
<div id="ref-YuJbg3zO" class="csl-entry" role="doc-biblioentry">
Himmelstein, D. S., Rubinetti, V., Slochower, D. R., Hu, D., Malladi, V. S., Greene, C. S., &amp; Gitter, A. (2019). Open collaborative writing with Manubot. <em>PLOS Computational Biology</em>, <em>15</em>(6), e1007128. <a href="https://doi.org/10.1371/journal.pcbi.1007128">https://doi.org/10.1371/journal.pcbi.1007128</a>
</div>
<div id="ref-YeFSbfFV" class="csl-entry" role="doc-biblioentry">
<em>Homepage | Travis CI – Start building today!</em> (n.d.). Travis CI. Retrieved November 17, 2022, from <a href="https://www.travis-ci.com/">https://www.travis-ci.com/</a>
</div>
<div id="ref-GQj3c17f" class="csl-entry" role="doc-biblioentry">
Hudgins, E. (2022). <em>emmajhudgins/UStreedamage: publication release</em> (accepted) [Computer software]. Zenodo. <a href="https://doi.org/10.5281/zenodo.6097109">https://doi.org/10.5281/zenodo.6097109</a>
</div>
<div id="ref-3UAritXO" class="csl-entry" role="doc-biblioentry">
<em>Issues · fmsabatini/sPlotOpen_Manuscript</em>. (n.d.). GitHub. Retrieved November 17, 2022, from <a href="https://github.com/fmsabatini/sPlotOpen_Manuscript">https://github.com/fmsabatini/sPlotOpen_Manuscript</a>
</div>
<div id="ref-wwHxTOtm" class="csl-entry" role="doc-biblioentry">
Kanza, S., Willoughby, C., Gibbins, N., Whitby, R., Frey, J. G., Erjavec, J., Zupančič, K., Hren, M., &amp; Kovač, K. (2017). Electronic lab notebooks: can they replace paper? <em>Journal of Cheminformatics</em>, <em>9</em>(1). <a href="https://doi.org/10.1186/s13321-017-0221-3">https://doi.org/10.1186/s13321-017-0221-3</a>
</div>
<div id="ref-cfgXxgt1" class="csl-entry" role="doc-biblioentry">
Keitt, T. H., &amp; Abelson, E. S. (2021). Ecology in the age of automation. <em>Science</em>, <em>373</em>(6557), 858–859. <a href="https://doi.org/10.1126/science.abi4692">https://doi.org/10.1126/science.abi4692</a>
</div>
<div id="ref-cW7vGddM" class="csl-entry" role="doc-biblioentry">
Khelifa, R., Amano, T., &amp; Nuñez, M. A. (2022). A solution for breaking the language barrier. <em>Trends in Ecology &amp;Amp; Evolution</em>, <em>37</em>(2), 109–112. <a href="https://doi.org/10.1016/j.tree.2021.11.003">https://doi.org/10.1016/j.tree.2021.11.003</a>
</div>
<div id="ref-lJAgyhYq" class="csl-entry" role="doc-biblioentry">
Kim, A. Y., Herrmann, V., Barreto, R., Calkins, B., Gonzalez‐Akre, E., Johnson, D. J., Jordan, J. A., Magee, L., McGregor, I. R., Montero, N., Novak, K., Rogers, T., Shue, J., &amp; Anderson‐Teixeira, K. J. (2022). Implementing GitHub Actions continuous integration to reduce error rates in ecological data collection. <em>Methods in Ecology and Evolution</em>, <em>13</em>(11), 2572–2585. <a href="https://doi.org/10.1111/2041-210x.13982">https://doi.org/10.1111/2041-210x.13982</a>
</div>
<div id="ref-139b0pSGc" class="csl-entry" role="doc-biblioentry">
Leibzon, W. (2016, August). Social network of software development at GitHub. <em>2016 IEEE/ACM International Conference on Advances in Social Networks Analysis and Mining (ASONAM)</em>. 2016 IEEE/ACM International Conference on Advances in Social Networks Analysis and Mining (ASONAM). <a href="https://doi.org/10.1109/asonam.2016.7752419">https://doi.org/10.1109/asonam.2016.7752419</a>
</div>
<div id="ref-3DKwn1sY" class="csl-entry" role="doc-biblioentry">
Lowndes, J. S. S., Best, B. D., Scarborough, C., Afflerbach, J. C., Frazier, M. R., O’Hara, C. C., Jiang, N., &amp; Halpern, B. S. (2017). Our path to better science in less time using open data science tools. <em>Nature Ecology &amp;Amp; Evolution</em>, <em>1</em>(6). <a href="https://doi.org/10.1038/s41559-017-0160">https://doi.org/10.1038/s41559-017-0160</a>
</div>
<div id="ref-ZI1OqZNr" class="csl-entry" role="doc-biblioentry">
Lrharper1. (2018). <em>Hulluni-Bioinformatics/Harper_Et_Al_2018: Github Repository Deposited With Harper Et Al. (2018) In Ecology And Evolution.</em> Zenodo. <a href="https://doi.org/10.5281/zenodo.1188710">https://doi.org/10.5281/zenodo.1188710</a>
</div>
<div id="ref-pjy75gHr" class="csl-entry" role="doc-biblioentry">
Madicken Munk, Koziar, K., Leinweber, K., Raniere Silva, Michonneau, F., McCue, R., Hejazi, N., Waldman, S., Emonet, R., Harris, R. M., Olex, A. L., Becker, E. A., Lever, J., Burle, M.-H., Moore, B., Umihiko Hoshijima, Maji, A., Topçuoğlu, B. D., Junghans, C., … Wolmar Nyberg Åkerström. (2019). <em>swcarpentry/git-novice: Software Carpentry: Version Control with Git, June 2019</em> (Version v2019.06.1). Zenodo. <a href="https://doi.org/10.5281/zenodo.3264950">https://doi.org/10.5281/zenodo.3264950</a>
</div>
<div id="ref-Re6Eg2va" class="csl-entry" role="doc-biblioentry">
<em>Manubot - Manuscripts, open and automated</em>. (n.d.). Retrieved November 17, 2022, from <a href="https://manubot.org">https://manubot.org</a>
</div>
<div id="ref-MwwMapRG" class="csl-entry" role="doc-biblioentry">
Marwick, B., Boettiger, C., &amp; Mullen, L. (2018). Packaging Data Analytical Work Reproducibly Using R (and Friends). <em>The American Statistician</em>, <em>72</em>(1), 80–88. <a href="https://doi.org/10.1080/00031305.2017.1375986">https://doi.org/10.1080/00031305.2017.1375986</a>
</div>
<div id="ref-mmCOSRfr" class="csl-entry" role="doc-biblioentry">
McIntire, E. J. B., Chubaty, A. M., Cumming, S. G., Andison, D., Barros, C., Boisvenue, C., Haché, S., Luo, Y., Micheletti, T., &amp; Stewart, F. E. C. (2022). PERFICT: A Re‐imagined foundation for predictive ecology. <em>Ecology Letters</em>, <em>25</em>(6), 1345–1351. <a href="https://doi.org/10.1111/ele.13994">https://doi.org/10.1111/ele.13994</a>
</div>
<div id="ref-1Ep9EJL6y" class="csl-entry" role="doc-biblioentry">
Meyer, M. (2014). Continuous Integration and Its Tools. <em>IEEE Software</em>, <em>31</em>(3), 14–16. <a href="https://doi.org/10.1109/ms.2014.58">https://doi.org/10.1109/ms.2014.58</a>
</div>
<div id="ref-bhDgD6lF" class="csl-entry" role="doc-biblioentry">
Micheletti, T., Stewart, F. E. C., Cumming, S. G., Haché, S., Stralberg, D., Tremblay, J. A., Barros, C., Eddy, I. M. S., Chubaty, A. M., Leblond, M., Pankratz, R. F., Mahon, C. L., Van Wilgenburg, S. L., Bayne, E. M., Schmiegelow, F., &amp; McIntire, E. J. B. (2021). Assessing Pathways of Climate Change Effects in SpaDES: An Application to Boreal Landbirds of Northwest Territories Canada. <em>Frontiers in Ecology and Evolution</em>, <em>9</em>. <a href="https://doi.org/10.3389/fevo.2021.679673">https://doi.org/10.3389/fevo.2021.679673</a>
</div>
<div id="ref-uBJwnPbq" class="csl-entry" role="doc-biblioentry">
Mislan, K. A. S., Heer, J. M., &amp; White, E. P. (2016). Elevating The Status of Code in Ecology. <em>Trends in Ecology &amp;Amp; Evolution</em>, <em>31</em>(1), 4–7. <a href="https://doi.org/10.1016/j.tree.2015.11.006">https://doi.org/10.1016/j.tree.2015.11.006</a>
</div>
<div id="ref-kZzfmBNu" class="csl-entry" role="doc-biblioentry">
Naujokaitis-Lewis, I., Endicott, S., &amp; Guezen, J. M. (2022). CAN-SAR: A database of Canadian species at risk information. <em>Scientific Data</em>, <em>9</em>(1). <a href="https://doi.org/10.1038/s41597-022-01381-8">https://doi.org/10.1038/s41597-022-01381-8</a>
</div>
<div id="ref-1Hcf13Q0k" class="csl-entry" role="doc-biblioentry">
Nugroho, R. P., Zuiderwijk, A., Janssen, M., &amp; de Jong, M. (2015). A comparison of national open data policies: lessons learned. <em>Transforming Government: People, Process and Policy</em>, <em>9</em>(3), 286–308. <a href="https://doi.org/10.1108/tg-03-2014-0008">https://doi.org/10.1108/tg-03-2014-0008</a>
</div>
<div id="ref-4LaijDIZ" class="csl-entry" role="doc-biblioentry">
On data availability, reproducibility and reuse. (2017). <em>Nature Cell Biology</em>, <em>19</em>(4), 259–259. <a href="https://doi.org/10.1038/ncb3506">https://doi.org/10.1038/ncb3506</a>
</div>
<div id="ref-kEX5dgzK" class="csl-entry" role="doc-biblioentry">
Perez-Riverol, Y., Gatto, L., Wang, R., Sachsenberg, T., Uszkoreit, J., Leprevost, F. da V., Fufezan, C., Ternent, T., Eglen, S. J., Katz, D. S., Pollard, T. J., Konovalov, A., Flight, R. M., Blin, K., &amp; Vizcaíno, J. A. (2016). Ten Simple Rules for Taking Advantage of Git and GitHub. <em>PLOS Computational Biology</em>, <em>12</em>(7), e1004947. <a href="https://doi.org/10.1371/journal.pcbi.1004947">https://doi.org/10.1371/journal.pcbi.1004947</a>
</div>
<div id="ref-10ghgV3S8" class="csl-entry" role="doc-biblioentry">
Perkel, J. (2016). Democratic databases: science on GitHub. <em>Nature</em>, <em>538</em>(7623), 127–128. <a href="https://doi.org/10.1038/538127a">https://doi.org/10.1038/538127a</a>
</div>
<div id="ref-1Kqna6l2" class="csl-entry" role="doc-biblioentry">
Perkel, J. M. (2020). Challenge to scientists: does your ten-year-old code still run? <em>Nature</em>, <em>584</em>(7822), 656–658. <a href="https://doi.org/10.1038/d41586-020-02462-7">https://doi.org/10.1038/d41586-020-02462-7</a>
</div>
<div id="ref-1CcAUn3Lu" class="csl-entry" role="doc-biblioentry">
Piwowar, H. A., Day, R. S., &amp; Fridsma, D. B. (2007). Sharing Detailed Research Data Is Associated with Increased Citation Rate. <em>PLoS ONE</em>, <em>2</em>(3), e308. <a href="https://doi.org/10.1371/journal.pone.0000308">https://doi.org/10.1371/journal.pone.0000308</a>
</div>
<div id="ref-10SpoByIw" class="csl-entry" role="doc-biblioentry">
Prabhu, P., Zhang, Y., Ghosh, S., August, D. I., Huang, J., Beard, S., Kim, H., Oh, T., Jablin, T. B., Johnson, N. P., Zoufaly, M., Raman, A., Liu, F., &amp; Walker, D. (2011). A survey of the practice of computational science. <em>State of the Practice Reports on - SC '11</em>. State of the Practice Reports. <a href="https://doi.org/10.1145/2063348.2063374">https://doi.org/10.1145/2063348.2063374</a>
</div>
<div id="ref-666HppfO" class="csl-entry" role="doc-biblioentry">
Pronk, T. E., Wiersma, P. H., van Weerden, A., &amp; Schieving, F. (2015). A game theoretic analysis of research data sharing. <em>PeerJ</em>, <em>3</em>, e1242. <a href="https://doi.org/10.7717/peerj.1242">https://doi.org/10.7717/peerj.1242</a>
</div>
<div id="ref-4ny1onB0" class="csl-entry" role="doc-biblioentry">
Ram, K. (2013). Git can facilitate greater reproducibility and increased transparency in science. <em>Source Code for Biology and Medicine</em>, <em>8</em>(1). <a href="https://doi.org/10.1186/1751-0473-8-7">https://doi.org/10.1186/1751-0473-8-7</a>
</div>
<div id="ref-12103x16N" class="csl-entry" role="doc-biblioentry">
Scheller, R. M., Sturtevant, B. R., Gustafson, E. J., Ward, B. C., &amp; Mladenoff, D. J. (2010). Increasing the reliability of ecological models using modern software engineering techniques. <em>Frontiers in Ecology and the Environment</em>, <em>8</em>(5), 253–260. <a href="https://doi.org/10.1890/080141">https://doi.org/10.1890/080141</a>
</div>
<div id="ref-10V7x4H4l" class="csl-entry" role="doc-biblioentry">
Schnell, S. (2015). Ten Simple Rules for a Computational Biologist’s Laboratory Notebook. <em>PLOS Computational Biology</em>, <em>11</em>(9), e1004385. <a href="https://doi.org/10.1371/journal.pcbi.1004385">https://doi.org/10.1371/journal.pcbi.1004385</a>
</div>
<div id="ref-1GRWGlDWy" class="csl-entry" role="doc-biblioentry">
Slack. (n.d.). <em>GitHub for Slack</em>. Slack Help Center. Retrieved November 17, 2022, from <a href="https://slack.com/help/articles/232289568-GitHub-for-Slack">https://slack.com/help/articles/232289568-GitHub-for-Slack</a>
</div>
<div id="ref-HiIPSSHV" class="csl-entry" role="doc-biblioentry">
Smaglik, P. (2007). Creating better lab websites gives potential collaborators and recruiters a clearer window into your world. <em>Nature</em>, <em>447</em>(7142), 347–347. <a href="https://doi.org/10.1038/nj7142-347a">https://doi.org/10.1038/nj7142-347a</a>
</div>
<div id="ref-ufw0ZdnI" class="csl-entry" role="doc-biblioentry">
Soergel, D. A. W. (2015). Rampant software errors may undermine scientific results. <em>F1000Research</em>, <em>3</em>, 303. <a href="https://doi.org/10.12688/f1000research.5930.2">https://doi.org/10.12688/f1000research.5930.2</a>
</div>
<div id="ref-hm9PaCLD" class="csl-entry" role="doc-biblioentry">
Song, X., Goldstein, S. C., &amp; Sakr, M. (2020, June 2). Using Peer Code Review as an Educational Tool. <em>Proceedings of the 2020 ACM Conference on Innovation and Technology in Computer Science Education</em>. ITiCSE '20: Innovation and Technology in Computer Science Education. <a href="https://doi.org/10.1145/3341525.3387370">https://doi.org/10.1145/3341525.3387370</a>
</div>
<div id="ref-1CzUZwyU2" class="csl-entry" role="doc-biblioentry">
Tenopir, C., Dalton, E. D., Allard, S., Frame, M., Pjesivac, I., Birch, B., Pollock, D., &amp; Dorsett, K. (2015). Changes in Data Sharing and Data Reuse Practices and Perceptions among Scientists Worldwide. <em>PLOS ONE</em>, <em>10</em>(8), e0134826. <a href="https://doi.org/10.1371/journal.pone.0134826">https://doi.org/10.1371/journal.pone.0134826</a>
</div>
<div id="ref-PLmDFZrm" class="csl-entry" role="doc-biblioentry">
Tenopir, C., Rice, N. M., Allard, S., Baird, L., Borycz, J., Christian, L., Grant, B., Olendorf, R., &amp; Sandusky, R. J. (2020). Data sharing, management, use, and reuse: Practices and perceptions of scientists worldwide. <em>PLOS ONE</em>, <em>15</em>(3), e0229003. <a href="https://doi.org/10.1371/journal.pone.0229003">https://doi.org/10.1371/journal.pone.0229003</a>
</div>
<div id="ref-dqrFjoSb" class="csl-entry" role="doc-biblioentry">
Trujillo, G., &amp; Tanner, K. D. (2014). Considering the Role of Affect in Learning: Monitoring Students' Self-Efficacy, Sense of Belonging, and Science Identity. <em>CBE—Life Sciences Education</em>, <em>13</em>(1), 6–15. <a href="https://doi.org/10.1187/cbe.13-12-0241">https://doi.org/10.1187/cbe.13-12-0241</a>
</div>
<div id="ref-1BJcvyTmV" class="csl-entry" role="doc-biblioentry">
Vale, G., Schmid, A., Santos, A. R., de Almeida, E. S., &amp; Apel, S. (2019). On the relation between Github communication activity and merge conflicts. <em>Empirical Software Engineering</em>, <em>25</em>(1), 402–433. <a href="https://doi.org/10.1007/s10664-019-09774-x">https://doi.org/10.1007/s10664-019-09774-x</a>
</div>
<div id="ref-19kmNxiHc" class="csl-entry" role="doc-biblioentry">
Vines, Timothy H., Albert, Arianne Y. K., Andrew, Rose L., Débarre, F., Bock, Dan G., Franklin, Michelle T., Gilbert, Kimberly J., Moore, J.-S., Renaut, S., &amp; Rennison, Diana J. (2014). The Availability of Research Data Declines Rapidly with Article Age. <em>Current Biology</em>, <em>24</em>(1), 94–97. <a href="https://doi.org/10.1016/j.cub.2013.11.014">https://doi.org/10.1016/j.cub.2013.11.014</a>
</div>
<div id="ref-SirQKFIz" class="csl-entry" role="doc-biblioentry">
White, E. P., Yenni, G. M., Taylor, S. D., Christensen, E. M., Bledsoe, E. K., Simonis, J. L., &amp; Ernest, S. K. M. (2018). Developing an automated iterative near‐term forecasting system for an ecological study. <em>Methods in Ecology and Evolution</em>, <em>10</em>(3), 332–344. <a href="https://doi.org/10.1111/2041-210x.13104">https://doi.org/10.1111/2041-210x.13104</a>
</div>
<div id="ref-1Ch6LSHef" class="csl-entry" role="doc-biblioentry">
Wicherts, J. M., Bakker, M., &amp; Molenaar, D. (2011). Willingness to Share Research Data Is Related to the Strength of the Evidence and the Quality of Reporting of Statistical Results. <em>PLoS ONE</em>, <em>6</em>(11), e26828. <a href="https://doi.org/10.1371/journal.pone.0026828">https://doi.org/10.1371/journal.pone.0026828</a>
</div>
<div id="ref-O6UbstGG" class="csl-entry" role="doc-biblioentry">
Wilson, G., Aruliah, D. A., Brown, C. T., Chue Hong, N. P., Davis, M., Guy, R. T., Haddock, S. H. D., Huff, K. D., Mitchell, I. M., Plumbley, M. D., Waugh, B., White, E. P., &amp; Wilson, P. (2014). Best Practices for Scientific Computing. <em>PLoS Biology</em>, <em>12</em>(1), e1001745. <a href="https://doi.org/10.1371/journal.pbio.1001745">https://doi.org/10.1371/journal.pbio.1001745</a>
</div>
<div id="ref-QqMezOMg" class="csl-entry" role="doc-biblioentry">
Xie, Y., Allaire, J. J., &amp; Grolemund, G. (2019). <em>R Markdown: the definitive guide</em>. CRC Press, Taylor and Francis Group.
</div>
<div id="ref-1CJo8lo2v" class="csl-entry" role="doc-biblioentry">
Yenni, G. M., Christensen, E. M., Bledsoe, E. K., Supp, S. R., Diaz, R. M., White, E. P., &amp; Ernest, S. K. M. (2019). Developing a modern data workflow for regularly updated data. <em>PLOS Biology</em>, <em>17</em>(1), e3000125. <a href="https://doi.org/10.1371/journal.pbio.3000125">https://doi.org/10.1371/journal.pbio.3000125</a>
</div>
<div id="ref-lXvpQxeN" class="csl-entry" role="doc-biblioentry">
Yenni, G. M., Christensen, E. M., Bledsoe, E. K., Supp, S. R., Diaz, R. M., White, E. P., &amp; Ernest, S. K. M. (2018). <em>Developing a modern data workflow for evolving data</em>. Cold Spring Harbor Laboratory. <a href="https://doi.org/10.1101/344804">https://doi.org/10.1101/344804</a>
</div>
</div>
<!-- default theme -->

<style>
  /* import google fonts */
  @import url("https://fonts.googleapis.com/css?family=Open+Sans:400,600,700");
  @import url("https://fonts.googleapis.com/css?family=Source+Code+Pro");

  /* -------------------------------------------------- */
  /* global */
  /* -------------------------------------------------- */

  /* all elements */
  * {
    /* force sans-serif font unless specified otherwise */
    font-family: "Open Sans", "Helvetica", sans-serif;

    /* prevent text inflation on some mobile browsers */
    -webkit-text-size-adjust: none !important;
    -moz-text-size-adjust: none !important;
    -o-text-size-adjust: none !important;
    text-size-adjust: none !important;
  }

  @media only screen {
    /* "page" element */
    body {
      position: relative;
      box-sizing: border-box;
      font-size: 12pt;
      line-height: 1.5;
      max-width: 8.5in;
      margin: 20px auto;
      padding: 40px;
      border-radius: 5px;
      border: solid 1px #bdbdbd;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      background: #ffffff;
    }
  }

  /* when on screen < 8.5in wide */
  @media only screen and (max-width: 8.5in) {
    /* "page" element */
    body {
      padding: 20px;
      margin: 0;
      border-radius: 0;
      border: none;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05) inset;
      background: none;
    }
  }

  /* -------------------------------------------------- */
  /* headings */
  /* -------------------------------------------------- */

  /* all headings */
  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    margin: 20px 0;
    padding: 0;
    font-weight: bold;
  }

  /* biggest heading */
  h1 {
    margin: 40px 0;
    text-align: center;
  }

  /* second biggest heading */
  h2 {
    margin-top: 30px;
    padding-bottom: 5px;
    border-bottom: solid 1px #bdbdbd;
  }

  /* heading font sizes */
  h1 {
    font-size: 2em;
  }
  h2 {
    font-size: 1.5em;
  }
  h3 {
    font-size: 1.35em;
  }
  h4 {
    font-size: 1.25em;
  }
  h5 {
    font-size: 1.15em;
  }
  h6 {
    font-size: 1em;
  }

  /* -------------------------------------------------- */
  /* manuscript header */
  /* -------------------------------------------------- */

  /* manuscript title */
  header > h1 {
    margin: 0;
  }

  /* manuscript title caption text (ie "automatically generated on") */
  header + p {
    text-align: center;
    margin-top: 10px;
  }

  /* -------------------------------------------------- */
  /* text elements */
  /* -------------------------------------------------- */

  /* links */
  a {
    color: #2196f3;
    overflow-wrap: break-word;
  }

  /* superscripts and subscripts */
  sub,
  sup {
    /* prevent from affecting line height */
    line-height: 0;
  }

  /* unordered and ordered lists*/
  ul,
  ol {
    padding-left: 20px;
  }

  /* class for styling text semibold */
  .semibold {
    font-weight: 600;
  }

  /* class for styling elements horizontally left aligned */
  .left {
    display: block;
    text-align: left;
    margin-left: auto;
    margin-right: 0;
    justify-content: left;
  }

  /* class for styling elements horizontally centered */
  .center {
    display: block;
    text-align: center;
    margin-left: auto;
    margin-right: auto;
    justify-content: center;
  }

  /* class for styling elements horizontally right aligned */
  .right {
    display: block;
    text-align: right;
    margin-left: 0;
    margin-right: auto;
    justify-content: right;
  }

  /* -------------------------------------------------- */
  /* section elements */
  /* -------------------------------------------------- */

  /* horizontal divider line */
  hr {
    border: none;
    height: 1px;
    background: #bdbdbd;
  }

  /* paragraphs, horizontal dividers, figures, tables, code */
  p,
  hr,
  figure,
  table,
  pre {
    /* treat all as "paragraphs", with consistent vertical margins */
    margin-top: 20px;
    margin-bottom: 20px;
  }

  /* -------------------------------------------------- */
  /* figures */
  /* -------------------------------------------------- */

  /* figure */
  figure {
    max-width: 100%;
    margin-left: auto;
    margin-right: auto;
  }

  /* figure caption */
  figcaption {
    padding: 0;
    padding-top: 10px;
  }

  /* figure image element */
  figure > img,
  figure > svg {
    max-width: 100%;
    display: block;
    margin-left: auto;
    margin-right: auto;
  }

  /* figure auto-number */
  img + figcaption > span:first-of-type,
  svg + figcaption > span:first-of-type {
    font-weight: bold;
    margin-right: 5px;
  }

  /* -------------------------------------------------- */
  /* tables */
  /* -------------------------------------------------- */

  /* table */
  table {
    border-collapse: collapse;
    border-spacing: 0;
    width: 100%;
    margin-left: auto;
    margin-right: auto;
  }

  /* table cells */
  th,
  td {
    border: solid 1px #bdbdbd;
    padding: 10px;
    /* squash table if too wide for page by forcing line breaks */
    overflow-wrap: break-word;
    word-break: break-word;
  }

  /* header row and even rows */
  th,
  tr:nth-child(2n) {
    background-color: #fafafa;
  }

  /* odd rows */
  tr:nth-child(2n + 1) {
    background-color: #ffffff;
  }

  /* table caption */
  caption {
    text-align: left;
    padding: 0;
    padding-bottom: 10px;
  }

  /* table auto-number */
  table > caption > span:first-of-type {
    font-weight: bold;
    margin-right: 5px;
  }

  /* -------------------------------------------------- */
  /* code */
  /* -------------------------------------------------- */

  /* multi-line code block */
  pre {
    padding: 10px;
    background-color: #eeeeee;
    color: #000000;
    border-radius: 5px;
    break-inside: avoid;
    text-align: left;
  }

  /* inline code, ie code within normal text */
  :not(pre) > code {
    padding: 0 4px;
    background-color: #eeeeee;
    color: #000000;
    border-radius: 5px;
  }

  /* code text */
  /* apply all children, to reach syntax highlighting sub-elements */
  code,
  code * {
    /* force monospace font */
    font-family: "Source Code Pro", "Courier New", monospace;
  }

  /* -------------------------------------------------- */
  /* quotes */
  /* -------------------------------------------------- */

  /* quoted text */
  blockquote {
    margin: 0;
    padding: 0;
    border-left: 4px solid #bdbdbd;
    padding-left: 16px;
    break-inside: avoid;
  }

  /* -------------------------------------------------- */
  /* banners */
  /* -------------------------------------------------- */

  /* info banners */
  .banner {
    box-sizing: border-box;
    display: block;
    position: relative;
    width: 100%;
    margin-top: 20px;
    margin-bottom: 20px;
    padding: 20px;
    text-align: center;
  }

  /* paragraph in banner */
  .banner > p {
    margin: 0;
  }

  /* -------------------------------------------------- */
  /* highlight colors */
  /* -------------------------------------------------- */

  .white {
    background: #ffffff;
  }
  .lightgrey {
    background: #eeeeee;
  }
  .grey {
    background: #757575;
  }
  .darkgrey {
    background: #424242;
  }
  .black {
    background: #000000;
  }
  .lightred {
    background: #ffcdd2;
  }
  .lightyellow {
    background: #ffecb3;
  }
  .lightgreen {
    background: #dcedc8;
  }
  .lightblue {
    background: #e3f2fd;
  }
  .lightpurple {
    background: #f3e5f5;
  }
  .red {
    background: #f44336;
  }
  .orange {
    background: #ff9800;
  }
  .yellow {
    background: #ffeb3b;
  }
  .green {
    background: #4caf50;
  }
  .blue {
    background: #2196f3;
  }
  .purple {
    background: #9c27b0;
  }
  .white,
  .lightgrey,
  .lightred,
  .lightyellow,
  .lightgreen,
  .lightblue,
  .lightpurple,
  .orange,
  .yellow,
  .white a,
  .lightgrey a,
  .lightred a,
  .lightyellow a,
  .lightgreen a,
  .lightblue a,
  .lightpurple a,
  .orange a,
  .yellow a {
    color: #000000;
  }
  .grey,
  .darkgrey,
  .black,
  .red,
  .green,
  .blue,
  .purple,
  .grey a,
  .darkgrey a,
  .black a,
  .red a,
  .green a,
  .blue a,
  .purple a {
    color: #ffffff;
  }

  /* -------------------------------------------------- */
  /* buttons */
  /* -------------------------------------------------- */

  /* class for styling links like buttons */
  .button {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    margin: 5px;
    padding: 10px 20px;
    font-size: 0.75em;
    font-weight: 600;
    text-transform: uppercase;
    text-decoration: none;
    letter-spacing: 1px;
    background: none;
    color: #2196f3;
    border: solid 1px #bdbdbd;
    border-radius: 5px;
  }

  /* buttons when hovered */
  .button:hover:not([disabled]),
  .icon_button:hover:not([disabled]) {
    cursor: pointer;
    background: #f5f5f5;
  }

  /* buttons when disabled */
  .button[disabled],
  .icon_button[disabled] {
    opacity: 0.35;
    pointer-events: none;
  }

  /* class for styling buttons containg only single icon */
  .icon_button {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    text-decoration: none;
    margin: 0;
    padding: 0;
    background: none;
    border-radius: 5px;
    border: none;
    width: 20px;
    height: 20px;
    min-width: 20px;
    min-height: 20px;
  }

  /* icon button inner svg image */
  .icon_button > svg {
    height: 16px;
  }

  /* -------------------------------------------------- */
  /* icons */
  /* -------------------------------------------------- */

  /* class for styling icons inline with text */
  .inline_icon {
    height: 1em;
    position: relative;
    top: 0.125em;
  }

  /* -------------------------------------------------- */
  /* references */
  /* -------------------------------------------------- */

  .csl-entry {
    margin-top: 15px;
    margin-bottom: 15px;
  }

  /* -------------------------------------------------- */
  /* print control */
  /* -------------------------------------------------- */

  @media print {
    @page {
      /* suggested printing margin */
      margin: 0.5in;
    }

    /* document and "page" elements */
    html,
    body {
      margin: 0;
      padding: 0;
      width: 100%;
      height: 100%;
    }

    /* "page" element */
    body {
      font-size: 11pt !important;
      line-height: 1.35;
    }

    /* all headings */
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
      margin: 15px 0;
    }

    /* figures and tables */
    figure,
    table {
      font-size: 0.85em;
    }

    /* table cells */
    th,
    td {
      padding: 5px;
    }

    /* shrink font awesome icons */
    i.fas,
    i.fab,
    i.far,
    i.fal {
      transform: scale(0.85);
    }

    /* decrease banner margins */
    .banner {
      margin-top: 15px;
      margin-bottom: 15px;
      padding: 15px;
    }

    /* class for centering an element vertically on its own page */
    .page_center {
      margin: auto;
      width: 100%;
      height: 100%;
      display: flex;
      align-items: center;
      vertical-align: middle;
      break-before: page;
      break-after: page;
    }

    /* always insert a page break before the element */
    .page_break_before {
      break-before: page;
    }

    /* always insert a page break after the element */
    .page_break_after {
      break-after: page;
    }

    /* avoid page break before the element */
    .page_break_before_avoid {
      break-before: avoid;
    }

    /* avoid page break after the element */
    .page_break_after_avoid {
      break-after: avoid;
    }

    /* avoid page break inside the element */
    .page_break_inside_avoid {
      break-inside: avoid;
    }
  }

  /* -------------------------------------------------- */
  /* override pandoc css quirks */
  /* -------------------------------------------------- */

  .sourceCode {
    /* prevent unsightly overflow in wide code blocks */
    overflow: auto !important;
  }

  div.sourceCode {
    /* prevent background fill on top-most code block  container */
    background: none !important;
  }

  .sourceCode * {
    /* force consistent line spacing */
    line-height: 1.5 !important;
  }

  div.sourceCode {
    /* style code block margins same as <pre> element */
    margin-top: 20px;
    margin-bottom: 20px;
  }

  /* -------------------------------------------------- */
  /* tablenos */
  /* -------------------------------------------------- */

  /* tablenos wrapper */
  .tablenos {
    width: 100%;
    margin: 20px 0;
  }

  .tablenos > table {
    /* move margins from table to table_wrapper to allow margin collapsing */
    margin: 0;
  }

  @media only screen {
    /* tablenos wrapper */
    .tablenos {
      /* show scrollbar on tables if necessary to prevent overflow */
      overflow-x: auto !important;
    }

    .tablenos th,
    .tablenos td {
      overflow-wrap: unset !important;
      word-break: unset !important;
    }

    /* table in wrapper */
    .tablenos table,
    .tablenos table * {
      /* don't break table words */
      overflow-wrap: normal !important;
    }
  }
</style>
<!-- 
    Plugin Core

    Functions needed for and shared across all first-party plugins.
-->

<script>
  // get element that is target of hash (from link element or url)
  function getHashTarget(link) {
    const hash = link ? link.hash : window.location.hash;
    const id = hash.slice(1);
    let target = document.querySelector(`[id="${id}"]`);
    if (!target) return;

    // if figure or table, modify target to get expected element
    if (id.indexOf("fig:") === 0) target = target.querySelector("figure");
    if (id.indexOf("tbl:") === 0) target = target.querySelector("table");

    return target;
  }

  // get position/dimensions of element or viewport
  function getRectInView(element) {
    let rect = {};
    rect.left = 0;
    rect.top = 0;
    rect.right = document.documentElement.clientWidth;
    rect.bottom = document.documentElement.clientHeight;
    let style = {};

    if (element instanceof HTMLElement) {
      rect = element.getBoundingClientRect();
      style = window.getComputedStyle(element);
    }

    const margin = {};
    margin.left = parseFloat(style.marginLeftWidth) || 0;
    margin.top = parseFloat(style.marginTopWidth) || 0;
    margin.right = parseFloat(style.marginRightWidth) || 0;
    margin.bottom = parseFloat(style.marginBottomWidth) || 0;

    const border = {};
    border.left = parseFloat(style.borderLeftWidth) || 0;
    border.top = parseFloat(style.borderTopWidth) || 0;
    border.right = parseFloat(style.borderRightWidth) || 0;
    border.bottom = parseFloat(style.borderBottomWidth) || 0;

    const newRect = {};
    newRect.left = rect.left + margin.left + border.left;
    newRect.top = rect.top + margin.top + border.top;
    newRect.right = rect.right + margin.right + border.right;
    newRect.bottom = rect.bottom + margin.bottom + border.bottom;
    newRect.width = newRect.right - newRect.left;
    newRect.height = newRect.bottom - newRect.top;

    return newRect;
  }

  // get position of element relative to page
  function getRectInPage(element) {
    const rect = getRectInView(element);
    const body = getRectInView(document.body);

    const newRect = {};
    newRect.left = rect.left - body.left;
    newRect.top = rect.top - body.top;
    newRect.right = rect.right - body.left;
    newRect.bottom = rect.bottom - body.top;
    newRect.width = rect.width;
    newRect.height = rect.height;

    return newRect;
  }

  // get closest element before specified element that matches query
  function firstBefore(element, query) {
    while (element && element !== document.body && !element.matches(query))
      element = element.previousElementSibling || element.parentNode;

    return element;
  }

  // check if element is part of collapsed heading
  function isCollapsed(element) {
    while (element && element !== document.body) {
      if (element.dataset.collapsed === "true") return true;
      element = element.parentNode;
    }
    return false;
  }

  // expand any collapsed parent containers of element if necessary
  function expandElement(element) {
    if (isCollapsed(element)) {
      // accordion plugin
      const heading = firstBefore(element, "h2");
      if (heading) heading.click();
      // details/summary HTML element
      const summary = firstBefore(element, "summary");
      if (summary) summary.click();
    }
  }

  // scroll to and focus element
  function goToElement(element, offset) {
    // expand accordion section if collapsed
    expandElement(element);
    const y =
      getRectInView(element).top -
      getRectInView(document.documentElement).top -
      (offset || 0);

    // trigger any function listening for "onscroll" event
    window.dispatchEvent(new Event("scroll"));
    window.scrollTo(0, y);
    document.activeElement.blur();
    element.focus();
  }

  // get list of elements after a start element up to element matching query
  function nextUntil(element, query, exclude) {
    const elements = [];
    while (((element = element.nextElementSibling), element)) {
      if (element.matches(query)) break;
      if (!element.matches(exclude)) elements.push(element);
    }
    return elements;
  }
</script>
<!--
  Accordion Plugin

  Allows sections of content under h2 headings to be collapsible.
-->

<script type="module">
  // whether to always start expanded ('false'), always start collapsed
  // ('true'), or start collapsed when screen small ('auto')
  const startCollapsed = "auto";

  // start script
  function start() {
    // run through each <h2> heading
    const headings = document.querySelectorAll("h2");
    for (const heading of headings) {
      addArrow(heading);

      // start expanded/collapsed based on option
      if (
        startCollapsed === "true" ||
        (startCollapsed === "auto" && isSmallScreen()) ||
        heading.dataset.collapsed === "true"
      )
        collapseHeading(heading);
      else expandElement(heading);
    }

    // attach hash change listener to window
    window.addEventListener("hashchange", onHashChange);
  }

  // when hash (eg manuscript.html#introduction) changes
  function onHashChange() {
    const target = getHashTarget();
    if (target) goToElement(target);
  }

  // add arrow to heading
  function addArrow(heading) {
    // add arrow button
    const arrow = document.createElement("button");
    arrow.innerHTML = document.querySelector(".icon_angle_down").innerHTML;
    arrow.classList.add("icon_button", "accordion_arrow");
    heading.insertBefore(arrow, heading.firstChild);

    // attach click listener to heading and button
    heading.addEventListener("click", onHeadingClick);
    arrow.addEventListener("click", onArrowClick);
  }

  // determine if on mobile-like device with small screen
  function isSmallScreen() {
    return Math.min(window.innerWidth, window.innerHeight) < 480;
  }

  // when <h2> heading is clicked
  function onHeadingClick(event) {
    // only collapse if <h2> itself is target of click (eg, user did
    // not click on anchor within <h2>)
    if (event.target === this) toggleCollapse(this);
  }

  // when arrow button is clicked
  function onArrowClick() {
    toggleCollapse(this.parentNode);
  }

  // collapse section if expanded, expand if collapsed
  function toggleCollapse(heading) {
    if (heading.dataset.collapsed === "false") collapseHeading(heading);
    else expandElement(heading);
  }

  // elements to exclude from collapse, such as table of contents panel,
  // hypothesis panel, etc
  const exclude = "#toc_panel, div.annotator-frame, #lightbox_overlay";

  // collapse section
  function collapseHeading(heading) {
    heading.setAttribute("data-collapsed", "true");
    const children = getChildren(heading);
    for (const child of children) child.setAttribute("data-collapsed", "true");
  }

  // expand section
  function expandElement(heading) {
    heading.setAttribute("data-collapsed", "false");
    const children = getChildren(heading);
    for (const child of children) child.setAttribute("data-collapsed", "false");
  }

  // get list of elements between this <h2> and next <h2> or <h1>
  // ("children" of the <h2> section)
  function getChildren(heading) {
    return nextUntil(heading, "h2, h1", exclude);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- angle down icon -->

<template class="icon_angle_down">
  <!-- modified from: https://fontawesome.com/icons/angle-down -->
  <svg width="16" height="16" viewBox="0 0 448 512">
    <path
      fill="currentColor"
      d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* accordion arrow button */
    .accordion_arrow {
      margin-right: 10px;
    }

    /* arrow icon when <h2> data-collapsed attribute true */
    h2[data-collapsed="true"] > .accordion_arrow > svg {
      transform: rotate(-90deg);
    }

    /* all elements (except <h2>'s) when data-collapsed attribute true */
    *:not(h2)[data-collapsed="true"] {
      display: none;
    }

    /* accordion arrow button when hovered and <h2>'s when hovered */
    .accordion_arrow:hover,
    h2[data-collapsed="true"]:hover,
    h2[data-collapsed="false"]:hover {
      cursor: pointer;
    }
  }

  /* always hide accordion arrow button on print */
  @media only print {
    .accordion_arrow {
      display: none;
    }
  }
</style>
<!--
  Anchors Plugin

  Adds an anchor next to each of a certain type of element that provides a
  human-readable url to that specific item/position in the document (e.g.
  "manuscript.html#abstract"). It also makes it such that scrolling out of view
  of a target removes its identifier from the url.
-->

<script type="module">
  // which types of elements to add anchors next to, in "document.querySelector"
  // format
  const typesQuery =
    'h1, h2, h3, div[id^="fig:"], div[id^="tbl:"], span[id^="eq:"]';

  // start script
  function start() {
    // add anchor to each element of specified types
    const elements = document.querySelectorAll(typesQuery);
    for (const element of elements) addAnchor(element);

    // attach scroll listener to window
    window.addEventListener("scroll", onScroll);
  }

  // when window is scrolled
  function onScroll() {
    // if url has hash and user has scrolled out of view of hash
    // target, remove hash from url
    const tolerance = 100;
    const target = getHashTarget();
    if (target) {
      if (
        target.getBoundingClientRect().top > window.innerHeight + tolerance ||
        target.getBoundingClientRect().bottom < 0 - tolerance
      )
        history.pushState(null, null, " ");
    }
  }

  // add anchor to element
  function addAnchor(element) {
    let addTo; // element to add anchor button to

    // if figure or table, modify withId and addTo to get expected
    // elements
    if (element.id.indexOf("fig:") === 0) {
      addTo = element.querySelector("figcaption");
    } else if (element.id.indexOf("tbl:") === 0) {
      addTo = element.querySelector("caption");
    } else if (element.id.indexOf("eq:") === 0) {
      addTo = element.querySelector(".eqnos-number");
    }

    addTo = addTo || element;
    const id = element.id || null;

    // do not add anchor if element doesn't have assigned id.
    // id is generated by pandoc and is assumed to be unique and
    // human-readable
    if (!id) return;

    // create anchor button
    const anchor = document.createElement("a");
    anchor.innerHTML = document.querySelector(".icon_link").innerHTML;
    anchor.title = "Link to this part of the document";
    anchor.classList.add("icon_button", "anchor");
    anchor.dataset.ignore = "true";
    anchor.href = "#" + id;
    addTo.appendChild(anchor);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- link icon -->

<template class="icon_link">
  <!-- modified from: https://fontawesome.com/icons/link -->
  <svg width="16" height="16" viewBox="0 0 512 512">
    <path
      fill="currentColor"
      d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* anchor button */
    .anchor {
      opacity: 0;
      margin-left: 5px;
    }

    /* anchor buttons within <h2>'s */
    h2 .anchor {
      margin-left: 10px;
    }

    /* anchor buttons when hovered/focused and anything containing an anchor button when hovered */
    *:hover > .anchor,
    .anchor:hover,
    .anchor:focus {
      opacity: 1;
    }

    /* anchor button when hovered */
    .anchor:hover {
      cursor: pointer;
    }
  }

  /* always show anchor button on devices with no mouse/hover ability */
  @media (hover: none) {
    .anchor {
      opacity: 1;
    }
  }

  /* always hide anchor button on print */
  @media only print {
    .anchor {
      display: none;
    }
  }
</style>
<!-- 
    Attributes Plugin

    Allows arbitrary HTML attributes to be attached to (almost) any element.
    Place an HTML comment inside or next to the desired element with the content:
    $attribute="value"
-->

<script type="module">
  // start script
  function start() {
    // get list of comments in document
    const comments = findComments();

    for (const comment of comments)
      if (comment.parentElement)
        addAttributes(comment.parentElement, comment.nodeValue.trim());
  }

  // add html attributes to specified element based on string of
  // html attributes and values
  function addAttributes(element, text) {
    // regex's for finding attribute/value pairs in the format of
    // attribute="value" or attribute='value
    const regex2 = /\$([a-zA-Z\-]+)?=\"(.+?)\"/;
    const regex1 = /\$([a-zA-Z\-]+)?=\'(.+?)\'/;

    // loop through attribute/value pairs
    let match;
    while ((match = text.match(regex2) || text.match(regex1))) {
      // get attribute and value from regex capture groups
      let attribute = match[1];
      let value = match[2];

      // remove from string
      text = text.substring(match.index + match[0].length);

      if (!attribute || !value) break;

      // set attribute of parent element
      try {
        element.setAttribute(attribute, value);
      } catch (error) {
        console.log(error);
      }

      // special case for colspan
      if (attribute === "colspan") removeTableCells(element, value);
    }
  }

  // get list of comment elements in document
  function findComments() {
    const comments = [];

    // iterate over comment nodes in document
    function acceptNode(node) {
      return NodeFilter.FILTER_ACCEPT;
    }
    const iterator = document.createNodeIterator(
      document.body,
      NodeFilter.SHOW_COMMENT,
      acceptNode
    );
    let node;
    while ((node = iterator.nextNode())) comments.push(node);

    return comments;
  }

  // remove certain number of cells after specified cell
  function removeTableCells(cell, number) {
    number = parseInt(number);
    if (!number) return;

    // remove elements
    for (; number > 1; number--) {
      if (cell.nextElementSibling) cell.nextElementSibling.remove();
    }
  }

  // start script on DOMContentLoaded instead of load to ensure this plugins
  // runs before other plugins
  window.addEventListener("DOMContentLoaded", start);
</script>
<!--
  Jump to First Plugin

  Adds a button next to each reference entry, figure, and table that jumps the
  page to the first occurrence of a link to that item in the manuscript.
-->

<script type="module">
  // whether to add buttons next to reference entries
  const references = "true";
  // whether to add buttons next to figures
  const figures = "true";
  // whether to add buttons next to tables
  const tables = "true";

  // start script
  function start() {
    if (references !== "false")
      makeButtons(`div[id^="ref-"]`, ".csl-left-margin", "reference");
    if (figures !== "false")
      makeButtons(`div[id^="fig:"]`, "figcaption", "figure");
    if (tables !== "false") makeButtons(`div[id^="tbl:"]`, "caption", "table");
  }

  // when jump button clicked
  function onButtonClick() {
    const first = getFirstOccurrence(this.dataset.id);
    if (!first) return;

    // update url hash so navigating "back" in history will return user to button
    window.location.hash = this.dataset.id;
    // scroll to link
    const timeout = function () {
      goToElement(first, window.innerHeight * 0.5);
    };
    window.setTimeout(timeout, 0);
  }

  // get first occurrence of link to item in document
  function getFirstOccurrence(id) {
    let query = "a";
    query += '[href="#' + id + '"]';
    // exclude buttons, anchor links, toc links, etc
    query += ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    return document.querySelector(query);
  }

  // add button next to each reference entry, figure, or table
  function makeButtons(query, containerQuery, subject) {
    const elements = document.querySelectorAll(query);
    for (const element of elements) {
      const id = element.id;
      const buttonContainer = element.querySelector(containerQuery);
      const first = getFirstOccurrence(id);

      // if can't find link to reference or place to put button, ignore
      if (!first || !buttonContainer) continue;

      // make jump button
      let button = document.createElement("button");
      button.classList.add("icon_button", "jump_arrow");
      button.title = `Jump to the first occurrence of this ${subject} in the document`;
      const icon = document.querySelector(".icon_angle_double_up");
      button.innerHTML = icon.innerHTML;
      button.dataset.id = id;
      button.dataset.ignore = "true";
      button.addEventListener("click", onButtonClick);
      buttonContainer.prepend(button);
    }
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- angle double up icon -->

<template class="icon_angle_double_up">
  <!-- modified from: https://fontawesome.com/icons/angle-double-up -->
  <svg width="16" height="16" viewBox="0 0 320 512">
    <path
      fill="currentColor"
      d="M177 255.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 351.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 425.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1zm-34-192L7 199.7c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l96.4-96.4 96.4 96.4c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9l-136-136c-9.2-9.4-24.4-9.4-33.8 0z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* jump button */
    .jump_arrow {
      position: relative;
      top: 0.125em;
      margin-right: 5px;
    }
  }

  /* always hide jump button on print */
  @media only print {
    .jump_arrow {
      display: none;
    }
  }
</style>
<!-- 
    Lightbox Plugin

    Makes it such that when a user clicks on an image, the image fills the
    screen and the user can pan/drag/zoom the image and navigate between other
    images in the document.
-->

<script type="module">
  // list of possible zoom/scale factors
  const zooms =
    "0.1, 0.25, 0.333333, 0.5, 0.666666, 0.75, 1, 1.25, 1.5, 1.75, 2, 2.5, 3, 3.5, 4, 5, 6, 7, 8";
  // whether to fit image to view ('fit'), display at 100% and shrink if
  // necessary ('shrink'), or always display at 100% ('100')
  const defaultZoom = "fit";
  // whether to zoom in/out toward center of view ('true') or mouse ('false')
  const centerZoom = "false";

  // start script
  function start() {
    // run through each <img> element
    const imgs = document.querySelectorAll("figure > img");
    let count = 1;
    for (const img of imgs) {
      img.classList.add("lightbox_document_img");
      img.dataset.number = count;
      img.dataset.total = imgs.length;
      img.addEventListener("click", openLightbox);
      count++;
    }

    // attach mouse and key listeners to window
    window.addEventListener("mousemove", onWindowMouseMove);
    window.addEventListener("keyup", onKeyUp);
  }

  // when mouse is moved anywhere in window
  function onWindowMouseMove(event) {
    window.mouseX = event.clientX;
    window.mouseY = event.clientY;
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    switch (event.key) {
      // trigger click of prev button
      case "ArrowLeft":
        const prevButton = document.getElementById("lightbox_prev_button");
        if (prevButton) prevButton.click();
        break;
      // trigger click of next button
      case "ArrowRight":
        const nextButton = document.getElementById("lightbox_next_button");
        if (nextButton) nextButton.click();
        break;
      // close on esc
      case "Escape":
        closeLightbox();
        break;
    }
  }

  // open lightbox
  function openLightbox() {
    const lightbox = makeLightbox(this);
    if (!lightbox) return;

    blurBody(lightbox);
    document.body.appendChild(lightbox);
  }

  // make lightbox
  function makeLightbox(img) {
    // delete lightbox if it exists, start fresh
    closeLightbox();

    // create screen overlay containing lightbox
    const overlay = document.createElement("div");
    overlay.id = "lightbox_overlay";

    // create image info boxes
    const numberInfo = document.createElement("div");
    const zoomInfo = document.createElement("div");
    numberInfo.id = "lightbox_number_info";
    zoomInfo.id = "lightbox_zoom_info";

    // create container for image
    const imageContainer = document.createElement("div");
    imageContainer.id = "lightbox_image_container";
    const lightboxImg = makeLightboxImg(
      img,
      imageContainer,
      numberInfo,
      zoomInfo
    );
    imageContainer.appendChild(lightboxImg);

    // create bottom container for caption and navigation buttons
    const bottomContainer = document.createElement("div");
    bottomContainer.id = "lightbox_bottom_container";
    const caption = makeCaption(img);
    const prevButton = makePrevButton(img);
    const nextButton = makeNextButton(img);
    bottomContainer.appendChild(prevButton);
    bottomContainer.appendChild(caption);
    bottomContainer.appendChild(nextButton);

    // attach top middle and bottom to overlay
    overlay.appendChild(numberInfo);
    overlay.appendChild(zoomInfo);
    overlay.appendChild(imageContainer);
    overlay.appendChild(bottomContainer);

    return overlay;
  }

  // make <img> object that is intuitively draggable and zoomable
  function makeLightboxImg(sourceImg, container, numberInfoBox, zoomInfoBox) {
    // create copy of source <img>
    const img = sourceImg.cloneNode(true);
    img.classList.remove("lightbox_document_img");
    img.removeAttribute("id");
    img.removeAttribute("width");
    img.removeAttribute("height");
    img.style.position = "unset";
    img.style.margin = "0";
    img.style.padding = "0";
    img.style.width = "";
    img.style.height = "";
    img.style.minWidth = "";
    img.style.minHeight = "";
    img.style.maxWidth = "";
    img.style.maxHeight = "";
    img.id = "lightbox_img";

    // build sorted list of zoomSteps
    const zoomSteps = zooms.split(/[^0-9.]/).map((step) => parseFloat(step));
    zoomSteps.sort((a, b) => a - b);

    // <img> object property variables
    let zoom = 1;
    let translateX = 0;
    let translateY = 0;
    let clickMouseX = undefined;
    let clickMouseY = undefined;
    let clickTranslateX = undefined;
    let clickTranslateY = undefined;

    updateNumberInfo();

    // update image numbers displayed in info box
    function updateNumberInfo() {
      numberInfoBox.innerHTML =
        sourceImg.dataset.number + " of " + sourceImg.dataset.total;
    }

    // update zoom displayed in info box
    function updateZoomInfo() {
      let zoomInfo = zoom * 100;
      if (!Number.isInteger(zoomInfo)) zoomInfo = zoomInfo.toFixed(2);
      zoomInfoBox.innerHTML = zoomInfo + "%";
    }

    // move to closest zoom step above current zoom
    const zoomIn = function () {
      for (const zoomStep of zoomSteps) {
        if (zoomStep > zoom) {
          zoom = zoomStep;
          break;
        }
      }
      updateTransform();
    };

    // move to closest zoom step above current zoom
    const zoomOut = function () {
      zoomSteps.reverse();
      for (const zoomStep of zoomSteps) {
        if (zoomStep < zoom) {
          zoom = zoomStep;
          break;
        }
      }
      zoomSteps.reverse();

      updateTransform();
    };

    // update display of <img> based on scale/translate properties
    const updateTransform = function () {
      // set transform
      img.style.transform =
        "translate(" +
        (translateX || 0) +
        "px," +
        (translateY || 0) +
        "px) scale(" +
        (zoom || 1) +
        ")";

      // get new width/height after scale
      const rect = img.getBoundingClientRect();
      // limit translate
      translateX = Math.max(translateX, -rect.width / 2);
      translateX = Math.min(translateX, rect.width / 2);
      translateY = Math.max(translateY, -rect.height / 2);
      translateY = Math.min(translateY, rect.height / 2);

      // set transform
      img.style.transform =
        "translate(" +
        (translateX || 0) +
        "px," +
        (translateY || 0) +
        "px) scale(" +
        (zoom || 1) +
        ")";

      updateZoomInfo();
    };

    // fit <img> to container
    const fit = function () {
      // no x/y offset, 100% zoom by default
      translateX = 0;
      translateY = 0;
      zoom = 1;

      // widths of <img> and container
      const imgWidth = img.naturalWidth;
      const imgHeight = img.naturalHeight;
      const containerWidth = parseFloat(
        window.getComputedStyle(container).width
      );
      const containerHeight = parseFloat(
        window.getComputedStyle(container).height
      );

      // how much zooming is needed to fit <img> to container
      const xRatio = imgWidth / containerWidth;
      const yRatio = imgHeight / containerHeight;
      const maxRatio = Math.max(xRatio, yRatio);
      const newZoom = 1 / maxRatio;

      // fit <img> to container according to option
      if (defaultZoom === "shrink") {
        if (maxRatio > 1) zoom = newZoom;
      } else if (defaultZoom === "fit") zoom = newZoom;

      updateTransform();
    };

    // when mouse wheel is rolled anywhere in container
    const onContainerWheel = function (event) {
      if (!event) return;

      // let ctrl + mouse wheel to zoom behave as normal
      if (event.ctrlKey) return;

      // prevent normal scroll behavior
      event.preventDefault();
      event.stopPropagation();

      // point around which to scale img
      const viewRect = container.getBoundingClientRect();
      const viewX = (viewRect.left + viewRect.right) / 2;
      const viewY = (viewRect.top + viewRect.bottom) / 2;
      const originX = centerZoom === "true" ? viewX : mouseX;
      const originY = centerZoom === "true" ? viewY : mouseY;

      // get point on image under origin
      const oldRect = img.getBoundingClientRect();
      const oldPercentX = (originX - oldRect.left) / oldRect.width;
      const oldPercentY = (originY - oldRect.top) / oldRect.height;

      // increment/decrement zoom
      if (event.deltaY < 0) zoomIn();
      if (event.deltaY > 0) zoomOut();

      // get offset between previous image point and origin
      const newRect = img.getBoundingClientRect();
      const offsetX = originX - (newRect.left + newRect.width * oldPercentX);
      const offsetY = originY - (newRect.top + newRect.height * oldPercentY);

      // translate image to keep image point under origin
      translateX += offsetX;
      translateY += offsetY;

      // perform translate
      updateTransform();
    };

    // when container is clicked
    function onContainerClick(event) {
      // if container itself is target of click, and not other
      // element above it
      if (event.target === this) closeLightbox();
    }

    // when mouse button is pressed on image
    const onImageMouseDown = function (event) {
      // store original mouse position relative to image
      clickMouseX = window.mouseX;
      clickMouseY = window.mouseY;
      clickTranslateX = translateX;
      clickTranslateY = translateY;
      event.stopPropagation();
      event.preventDefault();
    };

    // when mouse button is released anywhere in window
    const onWindowMouseUp = function (event) {
      // reset original mouse position
      clickMouseX = undefined;
      clickMouseY = undefined;
      clickTranslateX = undefined;
      clickTranslateY = undefined;

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("mouseup", onWindowMouseUp);
    };

    // when mouse is moved anywhere in window
    const onWindowMouseMove = function (event) {
      if (
        clickMouseX === undefined ||
        clickMouseY === undefined ||
        clickTranslateX === undefined ||
        clickTranslateY === undefined
      )
        return;

      // offset image based on original and current mouse position
      translateX = clickTranslateX + window.mouseX - clickMouseX;
      translateY = clickTranslateY + window.mouseY - clickMouseY;
      updateTransform();
      event.preventDefault();

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("mousemove", onWindowMouseMove);
    };

    // when window is resized
    const onWindowResize = function (event) {
      fit();

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("resize", onWindowResize);
    };

    // attach the necessary event listeners
    img.addEventListener("dblclick", fit);
    img.addEventListener("mousedown", onImageMouseDown);
    container.addEventListener("wheel", onContainerWheel);
    container.addEventListener("mousedown", onContainerClick);
    container.addEventListener("touchstart", onContainerClick);
    window.addEventListener("mouseup", onWindowMouseUp);
    window.addEventListener("mousemove", onWindowMouseMove);
    window.addEventListener("resize", onWindowResize);

    // run fit() after lightbox atttached to document and <img> Loaded
    // so needed container and img dimensions available
    img.addEventListener("load", fit);

    return img;
  }

  // make caption
  function makeCaption(img) {
    const caption = document.createElement("div");
    caption.id = "lightbox_caption";
    const captionSource = img.nextElementSibling;
    if (captionSource.tagName.toLowerCase() === "figcaption") {
      const captionCopy = makeCopy(captionSource);
      caption.innerHTML = captionCopy.innerHTML;
    }

    caption.addEventListener("touchstart", function (event) {
      event.stopPropagation();
    });

    return caption;
  }

  // make carbon copy of html dom element
  function makeCopy(source) {
    const sourceCopy = source.cloneNode(true);

    // delete elements marked with ignore (eg anchor and jump buttons)
    const deleteFromCopy = sourceCopy.querySelectorAll('[data-ignore="true"]');
    for (const element of deleteFromCopy) element.remove();

    // delete certain element attributes
    const attributes = [
      "id",
      "data-collapsed",
      "data-selected",
      "data-highlighted",
      "data-glow",
    ];
    for (const attribute of attributes) {
      sourceCopy.removeAttribute(attribute);
      const elements = sourceCopy.querySelectorAll("[" + attribute + "]");
      for (const element of elements) element.removeAttribute(attribute);
    }

    return sourceCopy;
  }

  // make button to jump to previous image in document
  function makePrevButton(img) {
    const prevButton = document.createElement("button");
    prevButton.id = "lightbox_prev_button";
    prevButton.title = "Jump to the previous image in the document [←]";
    prevButton.classList.add("icon_button", "lightbox_button");
    prevButton.innerHTML = document.querySelector(".icon_caret_left").innerHTML;

    // attach click listeners to button
    prevButton.addEventListener("click", function () {
      getPrevImg(img).click();
    });

    return prevButton;
  }

  // make button to jump to next image in document
  function makeNextButton(img) {
    const nextButton = document.createElement("button");
    nextButton.id = "lightbox_next_button";
    nextButton.title = "Jump to the next image in the document [→]";
    nextButton.classList.add("icon_button", "lightbox_button");
    nextButton.innerHTML = document.querySelector(
      ".icon_caret_right"
    ).innerHTML;

    // attach click listeners to button
    nextButton.addEventListener("click", function () {
      getNextImg(img).click();
    });

    return nextButton;
  }

  // get previous image in document
  function getPrevImg(img) {
    const imgs = document.querySelectorAll(".lightbox_document_img");

    // find index of provided img
    let index;
    for (index = 0; index < imgs.length; index++) {
      if (imgs[index] === img) break;
    }

    // wrap index to other side if < 1
    if (index - 1 >= 0) index--;
    else index = imgs.length - 1;
    return imgs[index];
  }

  // get next image in document
  function getNextImg(img) {
    const imgs = document.querySelectorAll(".lightbox_document_img");

    // find index of provided img
    let index;
    for (index = 0; index < imgs.length; index++) {
      if (imgs[index] === img) break;
    }

    // wrap index to other side if > total
    if (index + 1 <= imgs.length - 1) index++;
    else index = 0;
    return imgs[index];
  }

  // close lightbox
  function closeLightbox() {
    focusBody();

    const lightbox = document.getElementById("lightbox_overlay");
    if (lightbox) lightbox.remove();
  }

  // make all elements behind lightbox non-focusable
  function blurBody(overlay) {
    const all = document.querySelectorAll("*");
    for (const element of all) element.tabIndex = -1;
    document.body.classList.add("body_no_scroll");
  }

  // make all elements focusable again
  function focusBody() {
    const all = document.querySelectorAll("*");
    for (const element of all) element.removeAttribute("tabIndex");
    document.body.classList.remove("body_no_scroll");
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
  <!-- modified from: https://fontawesome.com/icons/caret-left -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
    ></path>
  </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
  <!-- modified from: https://fontawesome.com/icons/caret-right -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* regular <img> in document when hovered */
    img.lightbox_document_img:hover {
      cursor: pointer;
    }

    .body_no_scroll {
      overflow: hidden !important;
    }

    /* screen overlay */
    #lightbox_overlay {
      display: flex;
      flex-direction: column;
      position: fixed;
      left: 0;
      top: 0;
      right: 0;
      bottom: 0;
      background: rgba(0, 0, 0, 0.75);
      z-index: 3;
    }

    /* middle area containing lightbox image */
    #lightbox_image_container {
      flex-grow: 1;
      display: flex;
      justify-content: center;
      align-items: center;
      overflow: hidden;
      position: relative;
      padding: 20px;
    }

    /* bottom area containing caption */
    #lightbox_bottom_container {
      display: flex;
      justify-content: center;
      align-items: center;
      height: 100px;
      min-height: 100px;
      max-height: 100px;
      background: rgba(0, 0, 0, 0.5);
    }

    /* image number info text box */
    #lightbox_number_info {
      position: absolute;
      color: #ffffff;
      font-weight: 600;
      left: 2px;
      top: 0;
      z-index: 4;
    }

    /* zoom info text box */
    #lightbox_zoom_info {
      position: absolute;
      color: #ffffff;
      font-weight: 600;
      right: 2px;
      top: 0;
      z-index: 4;
    }

    /* copy of image caption */
    #lightbox_caption {
      box-sizing: border-box;
      display: inline-block;
      width: 100%;
      max-height: 100%;
      padding: 10px 0;
      text-align: center;
      overflow-y: auto;
      color: #ffffff;
    }

    /* navigation previous/next button */
    .lightbox_button {
      width: 100px;
      height: 100%;
      min-width: 100px;
      min-height: 100%;
      color: #ffffff;
    }

    /* navigation previous/next button when hovered */
    .lightbox_button:hover {
      background: none !important;
    }

    /* navigation button icon */
    .lightbox_button > svg {
      height: 25px;
    }

    /* figure auto-number */
    #lightbox_caption > span:first-of-type {
      font-weight: bold;
      margin-right: 5px;
    }

    /* lightbox image when hovered */
    #lightbox_img:hover {
      cursor: grab;
    }

    /* lightbox image when grabbed */
    #lightbox_img:active {
      cursor: grabbing;
    }
  }

  /* when on screen < 480px wide */
  @media only screen and (max-width: 480px) {
    /* make navigation buttons skinnier on small screens to make more room for caption text */
    .lightbox_button {
      width: 50px;
      min-width: 50px;
    }
  }

  /* always hide lightbox on print */
  @media only print {
    #lightbox_overlay {
      display: none;
    }
  }
</style>
<!-- 
  Link Highlight Plugin

  Makes it such that when a user hovers or focuses a link, other links that have
  the same target will be highlighted. It also makes it such that when clicking
  a link, the target of the link (eg reference, figure, table) is briefly
  highlighted.
-->

<script type="module">
  // whether to also highlight links that go to external urls
  const externalLinks = "false";
  // whether user must click off to unhighlight instead of just
  // un-hovering
  const clickUnhighlight = "false";
  // whether to also highlight links that are unique
  const highlightUnique = "true";

  // start script
  function start() {
    const links = getLinks();
    for (const link of links) {
      // attach mouse and focus listeners to link
      link.addEventListener("mouseenter", onLinkFocus);
      link.addEventListener("focus", onLinkFocus);
      link.addEventListener("mouseleave", onLinkUnhover);
    }

    // attach click and hash change listeners to window
    window.addEventListener("click", onClick);
    window.addEventListener("touchstart", onClick);
    window.addEventListener("hashchange", onHashChange);

    // run hash change on window load in case user has navigated
    // directly to hash
    onHashChange();
  }

  // when link is focused (tabbed to) or hovered
  function onLinkFocus() {
    highlight(this);
  }

  // when link is unhovered
  function onLinkUnhover() {
    if (clickUnhighlight !== "true") unhighlightAll();
  }

  // when the mouse is clicked anywhere in window
  function onClick(event) {
    unhighlightAll();
  }

  // when hash (eg manuscript.html#introduction) changes
  function onHashChange() {
    const target = getHashTarget();
    if (target) glowElement(target);
  }

  // start glow sequence on an element
  function glowElement(element) {
    const startGlow = function () {
      onGlowEnd();
      element.dataset.glow = "true";
      element.addEventListener("animationend", onGlowEnd);
    };
    const onGlowEnd = function () {
      element.removeAttribute("data-glow");
      element.removeEventListener("animationend", onGlowEnd);
    };
    startGlow();
  }

  // highlight link and all others with same target
  function highlight(link) {
    // force unhighlight all to start fresh
    unhighlightAll();

    // get links with same target
    if (!link) return;
    const sameLinks = getSameLinks(link);

    // if link unique and option is off, exit and don't highlight
    if (sameLinks.length <= 1 && highlightUnique !== "true") return;

    // highlight all same links, and "select" (special highlight) this
    // one
    for (const sameLink of sameLinks) {
      if (sameLink === link) sameLink.setAttribute("data-selected", "true");
      else sameLink.setAttribute("data-highlighted", "true");
    }
  }

  // unhighlight all links
  function unhighlightAll() {
    const links = getLinks();
    for (const link of links) {
      link.setAttribute("data-selected", "false");
      link.setAttribute("data-highlighted", "false");
    }
  }

  // get links with same target
  function getSameLinks(link) {
    const results = [];
    const links = getLinks();
    for (const otherLink of links) {
      if (otherLink.getAttribute("href") === link.getAttribute("href"))
        results.push(otherLink);
    }
    return results;
  }

  // get all links of types we wish to handle
  function getLinks() {
    let query = "a";
    if (externalLinks !== "true") query += '[href^="#"]';
    // exclude buttons, anchor links, toc links, etc
    query += ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    return document.querySelectorAll(query);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<style>
  @media only screen {
    /* anything with data-highlighted attribute true */
    [data-highlighted="true"] {
      background: #ffeb3b;
    }

    /* anything with data-selected attribute true */
    [data-selected="true"] {
      background: #ff8a65 !important;
    }

    /* animation definition for glow */
    @keyframes highlight_glow {
      0% {
        background: none;
      }
      10% {
        background: #bbdefb;
      }
      100% {
        background: none;
      }
    }

    /* anything with data-glow attribute true */
    [data-glow="true"] {
      animation: highlight_glow 2s;
    }
  }
</style>
<!--
  Table of Contents Plugin

  Provides a "table of contents" (toc) panel on the side of the document that
  allows the user to conveniently navigate between sections of the document.
-->

<script type="module">
  // which types of elements to add links for, in "document.querySelector" format
  const typesQuery = "h1, h2, h3";
  // whether toc starts open. use 'true' or 'false', or 'auto' to
  // use 'true' behavior when screen wide enough and 'false' when not
  const startOpen = "false";
  // whether toc closes when clicking on toc link. use 'true' or
  // 'false', or 'auto' to use 'false' behavior when screen wide
  // enough and 'true' when not
  const clickClose = "auto";
  // if list item is more than this many characters, text will be
  // truncated
  const charLimit = "50";
  // whether or not to show bullets next to each toc item
  const bullets = "false";

  // start script
  function start() {
    // make toc panel and populate with entries (links to document
    // sections)
    const panel = makePanel();
    if (!panel) return;
    makeEntries(panel);
    // attach panel to document after making entries, so 'toc' heading
    // in panel isn't included in toc
    document.body.insertBefore(panel, document.body.firstChild);

    // initial panel state
    if (startOpen === "true" || (startOpen === "auto" && !isSmallScreen()))
      openPanel();
    else closePanel();

    // attach click, scroll, and hash change listeners to window
    window.addEventListener("click", onClick);
    window.addEventListener("scroll", onScroll);
    window.addEventListener("hashchange", onScroll);
    window.addEventListener("keyup", onKeyUp);
    onScroll();

    // add class to push document body down out of way of toc button
    document.body.classList.add("toc_body_nudge");
  }

  // determine if screen wide enough to fit toc panel
  function isSmallScreen() {
    // in default theme:
    // 816px = 8.5in = width of "page" (<body>) element
    // 260px = min width of toc panel (*2 for both sides of <body>)
    return window.innerWidth < 816 + 260 * 2;
  }

  // when mouse is clicked anywhere in window
  function onClick() {
    if (isSmallScreen()) closePanel();
  }

  // when window is scrolled or hash changed
  function onScroll() {
    highlightViewed();
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    // close on esc
    if (event.key === "Escape") closePanel();
  }

  // find entry of currently viewed document section in toc and highlight
  function highlightViewed() {
    const firstId = getFirstInView(typesQuery);

    // get toc entries (links), unhighlight all, then highlight viewed
    const list = document.getElementById("toc_list");
    if (!firstId || !list) return;
    const links = list.querySelectorAll("a");
    for (const link of links) link.dataset.viewing = "false";
    const link = list.querySelector('a[href="#' + firstId + '"]');
    if (!link) return;
    link.dataset.viewing = "true";
  }

  // get first or previous toc listed element in top half of view
  function getFirstInView(query) {
    // get all elements matching query and with id
    const elements = document.querySelectorAll(query);
    const elementsWithIds = [];
    for (const element of elements) {
      if (element.id) elementsWithIds.push(element);
    }

    // get first or previous element in top half of view
    for (let i = 0; i < elementsWithIds.length; i++) {
      const element = elementsWithIds[i];
      const prevElement = elementsWithIds[Math.max(0, i - 1)];
      if (element.getBoundingClientRect().top >= 0) {
        if (element.getBoundingClientRect().top < window.innerHeight / 2)
          return element.id;
        else return prevElement.id;
      }
    }
  }

  // make panel
  function makePanel() {
    // create panel
    const panel = document.createElement("div");
    panel.id = "toc_panel";
    if (bullets === "true") panel.dataset.bullets = "true";

    // create header
    const header = document.createElement("div");
    header.id = "toc_header";

    // create toc button
    const button = document.createElement("button");
    button.id = "toc_button";
    button.innerHTML = document.querySelector(".icon_th_list").innerHTML;
    button.title = "Table of Contents";
    button.classList.add("icon_button");

    // create header text
    const text = document.createElement("h4");
    text.innerHTML = "Table of Contents";

    // create container for toc list
    const list = document.createElement("div");
    list.id = "toc_list";

    // attach click listeners
    panel.addEventListener("click", onPanelClick);
    header.addEventListener("click", onHeaderClick);
    button.addEventListener("click", onButtonClick);

    // attach elements
    header.appendChild(button);
    header.appendChild(text);
    panel.appendChild(header);
    panel.appendChild(list);

    return panel;
  }

  // create toc entries (links) to each element of the specified types
  function makeEntries(panel) {
    const elements = document.querySelectorAll(typesQuery);
    for (const element of elements) {
      // do not add link if element doesn't have assigned id
      if (!element.id) continue;

      // create link/list item
      const link = document.createElement("a");
      link.classList.add("toc_link");
      switch (element.tagName.toLowerCase()) {
        case "h1":
          link.dataset.level = "1";
          break;
        case "h2":
          link.dataset.level = "2";
          break;
        case "h3":
          link.dataset.level = "3";
          break;
        case "h4":
          link.dataset.level = "4";
          break;
      }
      link.title = element.innerText;
      let text = element.innerText;
      if (text.length > charLimit) text = text.slice(0, charLimit) + "...";
      link.innerHTML = text;
      link.href = "#" + element.id;
      link.addEventListener("click", onLinkClick);

      // attach link
      panel.querySelector("#toc_list").appendChild(link);
    }
  }

  // when panel is clicked
  function onPanelClick(event) {
    // stop click from propagating to window/document and closing panel
    event.stopPropagation();
  }

  // when header itself is clicked
  function onHeaderClick(event) {
    togglePanel();
  }

  // when button is clicked
  function onButtonClick(event) {
    togglePanel();
    // stop header underneath button from also being clicked
    event.stopPropagation();
  }

  // when link is clicked
  function onLinkClick(event) {
    if (clickClose === "true" || (clickClose === "auto" && isSmallScreen()))
      closePanel();
    else openPanel();
  }

  // open panel if closed, close if opened
  function togglePanel() {
    const panel = document.getElementById("toc_panel");
    if (!panel) return;

    if (panel.dataset.open === "true") closePanel();
    else openPanel();
  }

  // open panel
  function openPanel() {
    const panel = document.getElementById("toc_panel");
    if (panel) panel.dataset.open = "true";
  }

  // close panel
  function closePanel() {
    const panel = document.getElementById("toc_panel");
    if (panel) panel.dataset.open = "false";
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- th list icon -->

<template class="icon_th_list">
  <!-- modified from: https://fontawesome.com/icons/th-list -->
  <svg width="16" height="16" viewBox="0 0 512 512" tabindex="-1">
    <path
      fill="currentColor"
      d="M96 96c0 26.51-21.49 48-48 48S0 122.51 0 96s21.49-48 48-48 48 21.49 48 48zM48 208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm0 160c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm96-236h352c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
      tabindex="-1"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* toc panel */
    #toc_panel {
      box-sizing: border-box;
      position: fixed;
      top: 0;
      left: 0;
      background: #ffffff;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      z-index: 2;
    }

    /* toc panel when closed */
    #toc_panel[data-open="false"] {
      min-width: 60px;
      width: 60px;
      height: 60px;
      border-right: solid 1px #bdbdbd;
      border-bottom: solid 1px #bdbdbd;
    }

    /* toc panel when open */
    #toc_panel[data-open="true"] {
      min-width: 260px;
      max-width: 480px;
      /* keep panel edge consistent distance away from "page" edge */
      width: calc(((100vw - 8.5in) / 2) - 30px - 40px);
      bottom: 0;
      border-right: solid 1px #bdbdbd;
    }

    /* toc panel header */
    #toc_header {
      box-sizing: border-box;
      display: flex;
      flex-direction: row;
      align-items: center;
      height: 60px;
      margin: 0;
      padding: 20px;
    }

    /* toc panel header when hovered */
    #toc_header:hover {
      cursor: pointer;
    }

    /* toc panel header when panel open */
    #toc_panel[data-open="true"] > #toc_header {
      border-bottom: solid 1px #bdbdbd;
    }

    /* toc open/close header button */
    #toc_button {
      margin-right: 20px;
    }

    /* hide toc list and header text when closed */
    #toc_panel[data-open="false"] > #toc_header > *:not(#toc_button),
    #toc_panel[data-open="false"] > #toc_list {
      display: none;
    }

    /* toc list of entries */
    #toc_list {
      box-sizing: border-box;
      width: 100%;
      padding: 20px;
      position: absolute;
      top: calc(60px + 1px);
      bottom: 0;
      overflow: auto;
    }

    /* toc entry, link to section in document */
    .toc_link {
      display: block;
      padding: 5px;
      position: relative;
      font-weight: 600;
      text-decoration: none;
    }

    /* toc entry when hovered or when "viewed" */
    .toc_link:hover,
    .toc_link[data-viewing="true"] {
      background: #f5f5f5;
    }

    /* toc entry, level 1 indentation */
    .toc_link[data-level="1"] {
      margin-left: 0;
    }

    /* toc entry, level 2 indentation */
    .toc_link[data-level="2"] {
      margin-left: 20px;
    }

    /* toc entry, level 3 indentation */
    .toc_link[data-level="3"] {
      margin-left: 40px;
    }

    /* toc entry, level 4 indentation */
    .toc_link[data-level="4"] {
      margin-left: 60px;
    }

    /* toc entry bullets */
    #toc_panel[data-bullets="true"] .toc_link[data-level]:before {
      position: absolute;
      left: -15px;
      top: -1px;
      font-size: 1.5em;
    }

    /* toc entry, level 2 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="2"]:before {
      content: "\2022";
    }

    /* toc entry, level 3 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="3"]:before {
      content: "\25AB";
    }

    /* toc entry, level 4 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="4"]:before {
      content: "-";
    }
  }

  /* when on screen < 8.5in wide */
  @media only screen and (max-width: 8.5in) {
    /* push <body> ("page") element down to make room for toc icon */
    .toc_body_nudge {
      padding-top: 60px;
    }

    /* toc icon when panel closed and not hovered */
    #toc_panel[data-open="false"]:not(:hover) {
      background: rgba(255, 255, 255, 0.75);
    }
  }

  /* always hide toc panel on print */
  @media only print {
    #toc_panel {
      display: none;
    }
  }
</style>
<!-- 
  Tooltips Plugin

  Makes it such that when the user hovers or focuses a link to a citation or
  figure, a tooltip appears with a preview of the reference content, along with
  arrows to navigate between instances of the same reference in the document.
-->

<script type="module">
  // whether user must click off to close tooltip instead of just un-hovering
  const clickClose = "false";
  // delay (in ms) between opening and closing tooltip
  const delay = "100";

  // start script
  function start() {
    const links = getLinks();
    for (const link of links) {
      // attach hover and focus listeners to link
      link.addEventListener("mouseover", onLinkHover);
      link.addEventListener("mouseleave", onLinkUnhover);
      link.addEventListener("focus", onLinkFocus);
      link.addEventListener("touchend", onLinkTouch);
    }

    // attach mouse, key, and resize listeners to window
    window.addEventListener("mousedown", onClick);
    window.addEventListener("touchstart", onClick);
    window.addEventListener("keyup", onKeyUp);
    window.addEventListener("resize", onResize);
  }

  // when link is hovered
  function onLinkHover() {
    // function to open tooltip
    const delayOpenTooltip = function () {
      openTooltip(this);
    }.bind(this);

    // run open function after delay
    this.openTooltipTimer = window.setTimeout(delayOpenTooltip, delay);
  }

  // when mouse leaves link
  function onLinkUnhover() {
    // cancel opening tooltip
    window.clearTimeout(this.openTooltipTimer);

    // don't close on unhover if option specifies
    if (clickClose === "true") return;

    // function to close tooltip
    const delayCloseTooltip = function () {
      // if tooltip open and if mouse isn't over tooltip, close
      const tooltip = document.getElementById("tooltip");
      if (tooltip && !tooltip.matches(":hover")) closeTooltip();
    };

    // run close function after delay
    this.closeTooltipTimer = window.setTimeout(delayCloseTooltip, delay);
  }

  // when link is focused (tabbed to)
  function onLinkFocus(event) {
    openTooltip(this);
  }

  // when link is touched on touch screen
  function onLinkTouch(event) {
    // attempt to force hover state on first tap always, and trigger
    // regular link click (and navigation) on second tap
    if (event.target === document.activeElement) event.target.click();
    else {
      document.activeElement.blur();
      event.target.focus();
    }
    if (event.cancelable) event.preventDefault();
    event.stopPropagation();
    return false;
  }

  // when mouse is clicked anywhere in window
  function onClick(event) {
    closeTooltip();
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    switch (event.key) {
      // trigger click of prev button
      case "ArrowLeft":
        const prevButton = document.getElementById("tooltip_prev_button");
        if (prevButton) prevButton.click();
        break;
      // trigger click of next button
      case "ArrowRight":
        const nextButton = document.getElementById("tooltip_next_button");
        if (nextButton) nextButton.click();
        break;
      // close on esc
      case "Escape":
        closeTooltip();
        break;
    }
  }

  // when window is resized or zoomed
  function onResize() {
    closeTooltip();
  }

  // get all links of types we wish to handle
  function getLinks() {
    const queries = [];
    // exclude buttons, anchor links, toc links, etc
    const exclude =
      ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    queries.push('a[href^="#ref-"]' + exclude); // citation links
    queries.push('a[href^="#fig:"]' + exclude); // figure links
    const query = queries.join(", ");
    return document.querySelectorAll(query);
  }

  // get links with same target, get index of link in set, get total
  // same links
  function getSameLinks(link) {
    const sameLinks = [];
    const links = getLinks();
    for (const otherLink of links) {
      if (otherLink.getAttribute("href") === link.getAttribute("href"))
        sameLinks.push(otherLink);
    }

    return {
      elements: sameLinks,
      index: sameLinks.indexOf(link),
      total: sameLinks.length,
    };
  }

  // open tooltip
  function openTooltip(link) {
    // delete tooltip if it exists, start fresh
    closeTooltip();

    // make tooltip element
    const tooltip = makeTooltip(link);

    // if source couldn't be found and tooltip not made, exit
    if (!tooltip) return;

    // make navbar elements
    const navBar = makeNavBar(link);
    if (navBar) tooltip.firstElementChild.appendChild(navBar);

    // attach tooltip to page
    document.body.appendChild(tooltip);

    // position tooltip
    const position = function () {
      positionTooltip(link);
    };
    position();

    // if tooltip contains images, position again after they've loaded
    const imgs = tooltip.querySelectorAll("img");
    for (const img of imgs) img.addEventListener("load", position);
  }

  // close (delete) tooltip
  function closeTooltip() {
    const tooltip = document.getElementById("tooltip");
    if (tooltip) tooltip.remove();
  }

  // make tooltip
  function makeTooltip(link) {
    // get target element that link points to
    const source = getSource(link);

    // if source can't be found, exit
    if (!source) return;

    // create new tooltip
    const tooltip = document.createElement("div");
    tooltip.id = "tooltip";
    const tooltipContent = document.createElement("div");
    tooltipContent.id = "tooltip_content";
    tooltip.appendChild(tooltipContent);

    // make copy of source node and put in tooltip
    const sourceCopy = makeCopy(source);
    tooltipContent.appendChild(sourceCopy);

    // attach mouse event listeners
    tooltip.addEventListener("click", onTooltipClick);
    tooltip.addEventListener("mousedown", onTooltipClick);
    tooltip.addEventListener("touchstart", onTooltipClick);
    tooltip.addEventListener("mouseleave", onTooltipUnhover);

    // (for interaction with lightbox plugin)
    // transfer click on tooltip copied img to original img
    const sourceImg = source.querySelector("img");
    const sourceCopyImg = sourceCopy.querySelector("img");
    if (sourceImg && sourceCopyImg) {
      const clickImg = function () {
        sourceImg.click();
        closeTooltip();
      };
      sourceCopyImg.addEventListener("click", clickImg);
    }

    return tooltip;
  }

  // make carbon copy of html dom element
  function makeCopy(source) {
    const sourceCopy = source.cloneNode(true);

    // delete elements marked with ignore (eg anchor and jump buttons)
    const deleteFromCopy = sourceCopy.querySelectorAll('[data-ignore="true"]');
    for (const element of deleteFromCopy) element.remove();

    // delete certain element attributes
    const attributes = [
      "id",
      "data-collapsed",
      "data-selected",
      "data-highlighted",
      "data-glow",
      "class",
    ];
    for (const attribute of attributes) {
      sourceCopy.removeAttribute(attribute);
      const elements = sourceCopy.querySelectorAll("[" + attribute + "]");
      for (const element of elements) element.removeAttribute(attribute);
    }

    return sourceCopy;
  }

  // when tooltip is clicked
  function onTooltipClick(event) {
    // when user clicks on tooltip, stop click from transferring
    // outside of tooltip (eg, click off to close tooltip, or eg click
    // off to unhighlight same refs)
    event.stopPropagation();
  }

  // when tooltip is unhovered
  function onTooltipUnhover(event) {
    if (clickClose === "true") return;

    // make sure new mouse/touch/focus no longer over tooltip or any
    // element within it
    const tooltip = document.getElementById("tooltip");
    if (!tooltip) return;
    if (this.contains(event.relatedTarget)) return;

    closeTooltip();
  }

  // make nav bar to go betwen prev/next instances of same reference
  function makeNavBar(link) {
    // find other links to the same source
    const sameLinks = getSameLinks(link);

    // don't show nav bar when singular reference
    if (sameLinks.total <= 1) return;

    // find prev/next links with same target
    const prevLink = getPrevLink(link, sameLinks);
    const nextLink = getNextLink(link, sameLinks);

    // create nav bar
    const navBar = document.createElement("div");
    navBar.id = "tooltip_nav_bar";
    const text = sameLinks.index + 1 + " of " + sameLinks.total;

    // create nav bar prev/next buttons
    const prevButton = document.createElement("button");
    const nextButton = document.createElement("button");
    prevButton.id = "tooltip_prev_button";
    nextButton.id = "tooltip_next_button";
    prevButton.title =
      "Jump to the previous occurence of this item in the document [←]";
    nextButton.title =
      "Jump to the next occurence of this item in the document [→]";
    prevButton.classList.add("icon_button");
    nextButton.classList.add("icon_button");
    prevButton.innerHTML = document.querySelector(".icon_caret_left").innerHTML;
    nextButton.innerHTML =
      document.querySelector(".icon_caret_right").innerHTML;
    navBar.appendChild(prevButton);
    navBar.appendChild(document.createTextNode(text));
    navBar.appendChild(nextButton);

    // attach click listeners to buttons
    prevButton.addEventListener("click", function () {
      onPrevNextClick(link, prevLink);
    });
    nextButton.addEventListener("click", function () {
      onPrevNextClick(link, nextLink);
    });

    return navBar;
  }

  // get previous link with same target
  function getPrevLink(link, sameLinks) {
    if (!sameLinks) sameLinks = getSameLinks(link);
    // wrap index to other side if < 1
    let index;
    if (sameLinks.index - 1 >= 0) index = sameLinks.index - 1;
    else index = sameLinks.total - 1;
    return sameLinks.elements[index];
  }

  // get next link with same target
  function getNextLink(link, sameLinks) {
    if (!sameLinks) sameLinks = getSameLinks(link);
    // wrap index to other side if > total
    let index;
    if (sameLinks.index + 1 <= sameLinks.total - 1) index = sameLinks.index + 1;
    else index = 0;
    return sameLinks.elements[index];
  }

  // get element that is target of link or url hash
  function getSource(link) {
    const hash = link ? link.hash : window.location.hash;
    const id = hash.slice(1);
    let target = document.querySelector('[id="' + id + '"]');
    if (!target) return;

    // if ref or figure, modify target to get expected element
    if (id.indexOf("ref-") === 0) target = target.querySelector(":nth-child(2)");
    else if (id.indexOf("fig:") === 0) target = target.querySelector("figure");

    return target;
  }

  // when prev/next arrow button is clicked
  function onPrevNextClick(link, prevNextLink) {
    if (link && prevNextLink)
      goToElement(prevNextLink, window.innerHeight * 0.5);
  }

  // scroll to and focus element
  function goToElement(element, offset) {
    // expand accordion section if collapsed
    expandElement(element);
    const y =
      getRectInView(element).top -
      getRectInView(document.documentElement).top -
      (offset || 0);
    // trigger any function listening for "onscroll" event
    window.dispatchEvent(new Event("scroll"));
    window.scrollTo(0, y);
    document.activeElement.blur();
    element.focus();
  }

  // determine position to place tooltip based on link position in
  // viewport and tooltip size
  function positionTooltip(link, left, top) {
    const tooltipElement = document.getElementById("tooltip");
    if (!tooltipElement) return;

    // get convenient vars for position/dimensions of
    // link/tooltip/page/view
    link = getRectInPage(link);
    const tooltip = getRectInPage(tooltipElement);
    const view = getRectInPage();

    // horizontal positioning
    if (left)
      // use explicit value
      left = left;
    else if (link.left + tooltip.width < view.right)
      // fit tooltip to right of link
      left = link.left;
    else if (link.right - tooltip.width > view.left)
      // fit tooltip to left of link
      left = link.right - tooltip.width;
    // center tooltip in view
    else left = (view.right - view.left) / 2 - tooltip.width / 2;

    // vertical positioning
    if (top)
      // use explicit value
      top = top;
    else if (link.top - tooltip.height > view.top)
      // fit tooltip above link
      top = link.top - tooltip.height;
    else if (link.bottom + tooltip.height < view.bottom)
      // fit tooltip below link
      top = link.bottom;
    else {
      // center tooltip in view
      top = view.top + view.height / 2 - tooltip.height / 2;
      // nudge off of link to left/right if possible
      if (link.right + tooltip.width < view.right) left = link.right;
      else if (link.left - tooltip.width > view.left)
        left = link.left - tooltip.width;
    }

    tooltipElement.style.left = left + "px";
    tooltipElement.style.top = top + "px";
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
  <!-- modified from: https://fontawesome.com/icons/caret-left -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
    ></path>
  </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
  <!-- modified from: https://fontawesome.com/icons/caret-right -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* tooltip container */
    #tooltip {
      position: absolute;
      width: 50%;
      min-width: 240px;
      max-width: 75%;
      z-index: 1;
    }

    /* tooltip content */
    #tooltip_content {
      margin-bottom: 5px;
      padding: 20px;
      border-radius: 5px;
      border: solid 1px #bdbdbd;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      background: #ffffff;
      overflow-wrap: break-word;
    }

    /* tooltip copy of paragraphs and figures */
    #tooltip_content > p,
    #tooltip_content > figure {
      margin: 0;
      max-height: 320px;
      overflow-y: auto;
    }

    /* tooltip copy of <img> */
    #tooltip_content > figure > img,
    #tooltip_content > figure > svg {
      max-height: 260px;
    }

    /* navigation bar */
    #tooltip_nav_bar {
      margin-top: 10px;
      text-align: center;
    }

    /* navigation bar previous/next buton */
    #tooltip_nav_bar > .icon_button {
      position: relative;
      top: 3px;
    }

    /* navigation bar previous button */
    #tooltip_nav_bar > .icon_button:first-of-type {
      margin-right: 5px;
    }

    /* navigation bar next button */
    #tooltip_nav_bar > .icon_button:last-of-type {
      margin-left: 5px;
    }
  }

  /* always hide tooltip on print */
  @media only print {
    #tooltip {
      display: none;
    }
  }
</style>
<!--
  Analytics Plugin (third-party) 
  
  Copy and paste code from Google Analytics or similar service here.
-->
<!-- 
  Annotations Plugin

  Allows public annotation of the  manuscript. See https://web.hypothes.is/.
-->

<script type="module">
  // configuration
  window.hypothesisConfig = function () {
    return {
      branding: {
        accentColor: "#2196f3",
        appBackgroundColor: "#f8f8f8",
        ctaBackgroundColor: "#f8f8f8",
        ctaTextColor: "#000000",
        selectionFontFamily: "Open Sans, Helvetica, sans serif",
        annotationFontFamily: "Open Sans, Helvetica, sans serif",
      },
    };
  };

  // hypothesis client script
  const embed = "https://hypothes.is/embed.js";
  // hypothesis annotation count query url
  const query = "https://api.hypothes.is/api/search?limit=0&url=";

  // start script
  function start() {
    const button = makeButton();
    document.body.insertBefore(button, document.body.firstChild);
    insertCount(button);
  }

  // make button
  function makeButton() {
    // create button
    const button = document.createElement("button");
    button.id = "hypothesis_button";
    button.innerHTML = document.querySelector(".icon_hypothesis").innerHTML;
    button.title = "Hypothesis annotations";
    button.classList.add("icon_button");

    function onClick(event) {
      onButtonClick(event, button);
    }

    // attach click listeners
    button.addEventListener("click", onClick);

    return button;
  }

  // insert annotations count
  async function insertCount(button) {
    // get annotation count from Hypothesis based on url
    let count = "-";
    try {
      const canonical = document.querySelector('link[rel="canonical"]');
      const location = window.location;
      const url = encodeURIComponent((canonical || location).href);
      const response = await fetch(query + url);
      const json = await response.json();
      count = json.total || "-";
    } catch (error) {
      console.log(error);
    }

    // put count into button
    const counter = document.createElement("span");
    counter.id = "hypothesis_count";
    counter.innerHTML = count;
    button.title = "View " + count + " Hypothesis annotations";
    button.append(counter);
  }

  // when button is clicked
  function onButtonClick(event, button) {
    const script = document.createElement("script");
    script.src = embed;
    document.body.append(script);
    button.remove();
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- hypothesis icon -->

<template class="icon_hypothesis">
  <!-- modified from: https://simpleicons.org/icons/hypothesis.svg / https://git.io/Jf1VB -->
  <svg width="16" height="16" viewBox="0 0 24 24" tabindex="-1">
    <path
      fill="currentColor"
      d="M3.43 0C2.5 0 1.72 .768 1.72 1.72V18.86C1.72 19.8 2.5 20.57 3.43 20.57H9.38L12 24L14.62 20.57H20.57C21.5 20.57 22.29 19.8 22.29 18.86V1.72C22.29 .77 21.5 0 20.57 0H3.43M5.14 3.43H7.72V9.43S8.58 7.72 10.28 7.72C12 7.72 13.74 8.57 13.74 11.24V17.14H11.16V12C11.16 10.61 10.28 10.07 9.43 10.29C8.57 10.5 7.72 11.41 7.72 13.29V17.14H5.14V3.43M18 13.72C18.95 13.72 19.72 14.5 19.72 15.42A1.71 1.71 0 0 1 18 17.13A1.71 1.71 0 0 1 16.29 15.42C16.29 14.5 17.05 13.71 18 13.71Z"
      tabindex="-1"
    ></path>
  </svg>
</template>

<style>
  /* hypothesis activation button */
  #hypothesis_button {
    box-sizing: border-box;
    position: fixed;
    top: 0;
    right: 0;
    width: 60px;
    height: 60px;
    background: #ffffff;
    border-radius: 0;
    border-left: solid 1px #bdbdbd;
    border-bottom: solid 1px #bdbdbd;
    box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
    z-index: 2;
  }

  /* hypothesis button svg */
  #hypothesis_button > svg {
    position: relative;
    top: -4px;
  }

  /* hypothesis annotation count */
  #hypothesis_count {
    position: absolute;
    left: 0;
    right: 0;
    bottom: 5px;
  }

  /* side panel */
  .annotator-frame {
    width: 280px !important;
  }

  /* match highlight color to rest of theme */
  .annotator-highlights-always-on .annotator-hl {
    background-color: #ffeb3b !important;
  }

  /* match focused color to rest of theme */
  .annotator-hl.annotator-hl-focused {
    background-color: #ff8a65 !important;
  }

  /* match bucket bar color to rest of theme */
  .annotator-bucket-bar {
    background: #f5f5f5 !important;
  }

  /* always hide button, toolbar, and tooltip on print */
  @media only print {
    #hypothesis_button {
      display: none;
    }

    .annotator-frame {
      display: none !important;
    }

    hypothesis-adder {
      display: none !important;
    }
  }
</style>
<!-- 
  Mathjax Plugin (third-party) 

  Allows the proper rendering of math/equations written in LaTeX.
  See https://www.mathjax.org/.
-->

<script type="text/x-mathjax-config">
  // configuration
  MathJax.Hub.Config({
    "CommonHTML": { linebreaks: { automatic: true } },
    "HTML-CSS": { linebreaks: { automatic: true } },
    "SVG": { linebreaks: { automatic: true } },
    "fast-preview": { disabled: true }
  });
</script>

<script
  src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"
  integrity="sha512-0doc9hKxR3PYwso42RD1p5ySZpzzuDiOwMrdCEh2WdJZCjcmFKc/wEnL+z8fBQrnHoiNWbo+3fiGkOYXBdQp4A=="
  crossorigin="anonymous"
></script>

<style>
  /* mathjax containers */
  .math.display > span:not(.MathJax_Preview) {
    /* turn inline element (no dimensions) into block (allows fixed width and thus scrolling) */
    display: flex !important;
    overflow-x: auto !important;
    overflow-y: hidden !important;
    justify-content: center;
    align-items: center;
    margin: 0 !important;
  }

  /* right click menu */
  .MathJax_Menu {
    border-radius: 5px !important;
    border: solid 1px #bdbdbd !important;
    box-shadow: none !important;
  }

  /* equation auto-number */
  span[id^="eq:"] > span.math.display + span {
    font-weight: 600;
  }

  /* equation */
  span[id^="eq:"] > span.math.display > span {
    /* nudge to make room for equation auto-number and anchor */
    margin-right: 60px !important;
  }
</style>
</body>
</html>
